diff -c -r --new-file UICV3orig//Generic/src/Makefile.linux UICV3//Generic/src/Makefile.linux
*** UICV3orig//Generic/src/Makefile.linux
--- UICV3//Generic/src/Makefile.linux	Fri Aug  6 19:47:56 1993
***************
*** 0 ****
--- 1,51 ----
+ #
+ #      SCCS ID: @(#)Makefile.g++ 1.2 93/02/08 SMI
+ #
+ #      Copyright (c) 1990-1992 Sun Microsystems, Inc.  All rights reserved.
+ #      See LEGAL_NOTICE file for terms of the license.
+ #
+ CCFLAGS  = -c -O $(DLL)
+ #CCFLAGS  = -c -g
+ 
+ INCLUDE = ../include/
+ LIB     = ../lib/
+ TEST    = ../test/
+ 
+ OBJECTS = Generic.o GenericHash.o GenericList.o
+ 
+ all: $(LIB)libGeneric.a
+ 
+ $(LIB)libGeneric.a : $(OBJECTS)
+ 	ranlib $(LIB)libGeneric.a
+ 
+ xs:
+ 	#load $(CCFLAGS) -I$(INCLUDE) $(OBJECTS:.o=.cc)
+ 
+ Generic.o : Generic.cc $(INCLUDE)Generic.h
+ 	g++ $(CCFLAGS) Generic.cc -I$(INCLUDE)
+ 	ar rv $(LIB)libGeneric.a Generic.o
+ 
+ GenericHash.o : GenericHash.cc $(INCLUDE)GenericHash.h
+ 	g++ $(CCFLAGS) GenericHash.cc -I$(INCLUDE)
+ 	ar rv $(LIB)libGeneric.a GenericHash.o
+ 
+ GenericList.o : GenericList.cc $(INCLUDE)GenericList.h
+ 	g++ $(CCFLAGS) GenericList.cc -I$(INCLUDE)
+ 	ar rv $(LIB)libGeneric.a GenericList.o
+ 
+ delete:
+ 	/bin/rm -f *.o $(LIB)*.a $(TEST)*test
+ 
+ clean:
+ 	/bin/rm -f *.o ../lib/*.a
+ 
+ test: hashtest listtest
+ 
+ hashtest:
+ 	g++ -o hashtest GenericHash.cc -DTEST -L$(LIB) -lGeneric -lm \
+ 	       -I$(INCLUDE)
+ 	mv hashtest $(TEST)
+ 
+ listtest:
+ 	g++ -o listtest GenericList.cc -DTEST -L$(LIB) -lGeneric -I$(INCLUDE)
+ 	mv listtest $(TEST)
diff -c -r --new-file UICV3orig//Makefile UICV3//Makefile
*** UICV3orig//Makefile	Fri Jun 18 04:01:21 1993
--- UICV3//Makefile	Sat Aug  7 14:23:44 1993
***************
*** 30,37 ****
  	cd $(UITHOME)/src/UIC; make -f Makefile.g++
  
  world_g++_svr4:
! 	cd $(GENERICHOME)/src; make -f Makefile.g++.svr4
! 	cd $(UITHOME)/src/UIC; make -f Makefile.g++.svr4
  
  world_g++_NO_XVPS:
  	cd $(GENERICHOME)/src; make -f Makefile.g++
--- 30,37 ----
  	cd $(UITHOME)/src/UIC; make -f Makefile.g++
  
  world_g++_svr4:
! 	cd $(GENERICHOME)/src; make -f Makefile.svr4.g++
! 	cd $(UITHOME)/src/UIC; make -f Makefile.svr4.g++
  
  world_g++_NO_XVPS:
  	cd $(GENERICHOME)/src; make -f Makefile.g++
***************
*** 38,46 ****
  	cd $(UITHOME)/src/UIC; make -f Makefile.g++.NO_XVPS
  
  world_g++_svr4_NO_XVPS:
! 	cd $(GENERICHOME)/src; make -f Makefile.g++.svr4
  	cd $(UITHOME)/src/UIC; make -f Makefile.g++.svr4.NO_XVPS
  
  delete:
  	cd $(GENERICHOME)/src; make -i delete
  	cd $(UITHOME)/src/UIC; make -i delete
--- 38,93 ----
  	cd $(UITHOME)/src/UIC; make -f Makefile.g++.NO_XVPS
  
  world_g++_svr4_NO_XVPS:
! 	cd $(GENERICHOME)/src; make -f Makefile.svr4.g++
  	cd $(UITHOME)/src/UIC; make -f Makefile.g++.svr4.NO_XVPS
  
  delete:
  	cd $(GENERICHOME)/src; make -i delete
  	cd $(UITHOME)/src/UIC; make -i delete
+ 
+ world_linux:
+ 	cd $(GENERICHOME)/src; make -f Makefile.linux
+ 	cd $(UITHOME)/src/UIC; make -f Makefile.linux
+ 
+ install_linux:
+ 	cp $(UITHOME)/lib/*.a $(OPENWINHOME)/lib
+ 	ranlib $(OPENWINHOME)/lib/libUIC.a
+ 	ranlib $(OPENWINHOME)/lib/libUICSS.a
+ 	-mkdir $(OPENWINHOME)/include/uit
+ 	(cd $(UITHOME)/include/c++ && tar cf - * | (cd $(OPENWINHOME)/include/uit && tar xf -))
+ 
+ examples_linux:
+ 	(cd $(UITHOME)/src/examples/ColorMaps/Display && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/ColorMaps/Worms && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/DragAndDrop/Panel && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm")
+ 	(cd $(UITHOME)/src/examples/DragAndDrop/Canvas && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm")
+ 	(cd $(UITHOME)/src/examples/DragAndDrop/PrintTool && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm")
+ 	(cd $(UITHOME)/src/examples/HotRegions/Canvas/Xlib && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/Images/List && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/Images/Choice && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/Menus && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/NewUITClasses/PropertiesWindow && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/NotifyTargets && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/OverviewExample && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/Windows && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUIC -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/SlingShot/CSArrayTile && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUICSS -lUIC -lsspkg -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/SlingShot/CSDrawImage && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUICSS -lUIC -lsspkg -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 	(cd $(UITHOME)/src/examples/SlingShot/CSDrawText && make "C++FLAGS=-I$(OPENWINHOME)/include/uit -I$(OPENWINHOME)/include" "LDLIBS=-L$(OPENWINHOME)/lib -L/usr/X386/lib -lUICSS -lUIC -lsspkg -lxview -lolgx -lX11 -lm" -f Makefile.g++.NO_XVPS)
+ 
+ examples_delete:
+ 	(cd $(UITHOME)/src/examples/ColorMaps/Display && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/ColorMaps/Worms && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/DragAndDrop/Panel && make delete)
+ 	(cd $(UITHOME)/src/examples/DragAndDrop/Canvas && make delete)
+ 	(cd $(UITHOME)/src/examples/DragAndDrop/PrintTool && make delete)
+ 	(cd $(UITHOME)/src/examples/HotRegions/Canvas/Xlib && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/Images/List && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/Images/Choice && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/Menus && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/NewUITClasses/PropertiesWindow && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/NotifyTargets && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/OverviewExample && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/Windows && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/SlingShot/CSArrayTile && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/SlingShot/CSDrawImage && make -f Makefile.g++.NO_XVPS delete)
+ 	(cd $(UITHOME)/src/examples/SlingShot/CSDrawText && make -f Makefile.g++.NO_XVPS delete)
diff -c -r --new-file UICV3orig//README.linux UICV3//README.linux
*** UICV3orig//README.linux
--- UICV3//README.linux	Fri Aug  6 21:11:48 1993
***************
*** 0 ****
--- 1,17 ----
+ How to build the UIC and UICSS static libraries under linux:
+ 
+ 1) Source the 'envvars' script while sitting in the toplevel directory of
+    the patched UIT source tree (same directory as this file)
+ 
+ . envvars
+ 
+ 2) Build the libraries. Some warnings about illegal casts may be ignored
+    (you are of course welcome to fix them if you want to contribute).
+ 
+ make world_linux
+ 
+ 3) Install the libraries and header files in /usr/openwin.
+ 
+ make install_linux
+ 
+ lmfken@lmf.ericsson.se
diff -c -r --new-file UICV3orig//UIT/include/c++/ComponentDisplay.h UICV3//UIT/include/c++/ComponentDisplay.h
*** UICV3orig//UIT/include/c++/ComponentDisplay.h	Fri Jun 18 04:02:09 1993
--- UICV3//UIT/include/c++/ComponentDisplay.h	Sat Jun 26 15:11:46 1993
***************
*** 452,458 ****
  #include <xview/xv_xrect.h>
  #include <xview/xview.h>
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #include "private/postscript_canvas.h"
  #endif
--- 452,458 ----
  #include <xview/xv_xrect.h>
  #include <xview/xview.h>
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #include "private/postscript_canvas.h"
  #endif
***************
*** 459,465 ****
  #endif
  
  // Macro Definitions
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #define POSTSCRIPT_STRING(string) pprintf (PostScript, \
  					   (unsigned char *) string, \
--- 459,465 ----
  #endif
  
  // Macro Definitions
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #define POSTSCRIPT_STRING(string) pprintf (PostScript, \
  					   (unsigned char *) string, \
***************
*** 644,650 ****
      void		 setRepaintHandler (UICHandlerFunction);
      void		 setResizeHandler (UICHandlerFunction);
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      void		 setPSCanvasClipRects (Xv_xrectlist *rects)
  			 { if ((model == POSTSCRIPT) && !isPanel)
--- 644,650 ----
      void		 setRepaintHandler (UICHandlerFunction);
      void		 setResizeHandler (UICHandlerFunction);
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      void		 setPSCanvasClipRects (Xv_xrectlist *rects)
  			 { if ((model == POSTSCRIPT) && !isPanel)
***************
*** 661,667 ****
      Display             *getXDisplay () { return canvasXDisplay; }
      Window               getXWindow () { return canvasXWindow; }
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      NeWStoken	         getPSWindow ()
  			 {
--- 661,667 ----
      Display             *getXDisplay () { return canvasXDisplay; }
      Window               getXWindow () { return canvasXWindow; }
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      NeWStoken	         getPSWindow ()
  			 {
***************
*** 725,731 ****
  	if (free_gc) XFreeGC (display, gc);
        }
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      void                 setPSFocus ()
        {
--- 725,731 ----
  	if (free_gc) XFreeGC (display, gc);
        }
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      void                 setPSFocus ()
        {
***************
*** 886,892 ****
      UICHandlerFunction	 repaintHandler;
      UICHandlerFunction   resizeHandler;
      Xv_window            paintWindow;
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      NeWStoken		 canvasPSWindow;
  #endif
--- 886,892 ----
      UICHandlerFunction	 repaintHandler;
      UICHandlerFunction   resizeHandler;
      Xv_window            paintWindow;
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      NeWStoken		 canvasPSWindow;
  #endif
***************
*** 929,935 ****
  			       Display *, 
  			       Window,
  			       Xv_xrectlist *);
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      static  void repaintPSProc (PScanvas,
  				NeWStoken,
--- 929,935 ----
  			       Display *, 
  			       Window,
  			       Xv_xrectlist *);
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      static  void repaintPSProc (PScanvas,
  				NeWStoken,
diff -c -r --new-file UICV3orig//UIT/include/c++/UIObject.h UICV3//UIT/include/c++/UIObject.h
*** UICV3orig//UIT/include/c++/UIObject.h	Fri Jun 18 04:02:12 1993
--- UICV3//UIT/include/c++/UIObject.h	Sat Jun 26 15:09:15 1993
***************
*** 959,965 ****
  #include "GenericHash.h"
  #include "GenericList.h"
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #include "private/postscript_canvas.h"
  #endif
--- 959,965 ----
  #include "GenericHash.h"
  #include "GenericList.h"
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #include "private/postscript_canvas.h"
  #endif
diff -c -r --new-file UICV3orig//UIT/src/UIC/ChoiceMenu.cc UICV3//UIT/src/UIC/ChoiceMenu.cc
*** UICV3orig//UIT/src/UIC/ChoiceMenu.cc	Fri Jun 18 04:02:23 1993
--- UICV3//UIT/src/UIC/ChoiceMenu.cc	Tue Aug 10 16:43:41 1993
***************
*** 319,325 ****
      }
  
      if (!menu_item->active)
!       xv_set (xview_menu_item, MENU_INACTIVE, TRUE);
  
      xv_set (owObject, MENU_APPEND_ITEM, xview_menu_item, NULL);
    }
--- 319,325 ----
      }
  
      if (!menu_item->active)
!       xv_set (xview_menu_item, MENU_INACTIVE, TRUE, NULL);
  
      xv_set (owObject, MENU_APPEND_ITEM, xview_menu_item, NULL);
    }
diff -c -r --new-file UICV3orig//UIT/src/UIC/CommandMenu.cc UICV3//UIT/src/UIC/CommandMenu.cc
*** UICV3orig//UIT/src/UIC/CommandMenu.cc	Fri Jun 18 04:02:23 1993
--- UICV3//UIT/src/UIC/CommandMenu.cc	Tue Aug 10 16:40:39 1993
***************
*** 138,144 ****
      }
  
      if (!menu_item->active)
!       xv_set (xview_menu_item, MENU_INACTIVE, TRUE);
  
      xv_set (owObject, MENU_APPEND_ITEM, xview_menu_item, NULL);
    }
--- 138,144 ----
      }
  
      if (!menu_item->active)
!       xv_set (xview_menu_item, MENU_INACTIVE, TRUE, NULL);
  
      xv_set (owObject, MENU_APPEND_ITEM, xview_menu_item, NULL);
    }
diff -c -r --new-file UICV3orig//UIT/src/UIC/ComponentDisplay.cc UICV3//UIT/src/UIC/ComponentDisplay.cc
*** UICV3orig//UIT/src/UIC/ComponentDisplay.cc	Fri Jun 18 04:02:32 1993
--- UICV3//UIT/src/UIC/ComponentDisplay.cc	Sat Jun 26 15:14:51 1993
***************
*** 111,117 ****
    paintWindow = (Xv_window) NULL;
    canvasXDisplay = (Display *) NULL;
    canvasXWindow = (Window) NULL;
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
    canvasPSWindow = (NeWStoken) UIC_NOT_SET;
  #endif
--- 111,117 ----
    paintWindow = (Xv_window) NULL;
    canvasXDisplay = (Display *) NULL;
    canvasXWindow = (Window) NULL;
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
    canvasPSWindow = (NeWStoken) UIC_NOT_SET;
  #endif
***************
*** 525,531 ****
  		CANVAS_REPAINT_PROC, &ComponentDisplay::repaintXProc,
  		NULL);
      }
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      // Else, if using the PostScript drawing model...
      else if (model == POSTSCRIPT) {
--- 525,531 ----
  		CANVAS_REPAINT_PROC, &ComponentDisplay::repaintXProc,
  		NULL);
      }
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      // Else, if using the PostScript drawing model...
      else if (model == POSTSCRIPT) {
***************
*** 693,699 ****
    }
  }
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  /*F+ void ComponentDisplay::repaintPSProc (Canvas        canvas,
  *                                          NeWStoken     ps_window,
--- 693,699 ----
    }
  }
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  /*F+ void ComponentDisplay::repaintPSProc (Canvas        canvas,
  *                                          NeWStoken     ps_window,
***************
*** 1119,1125 ****
  		  CANVAS_REPAINT_PROC, &ComponentDisplay::repaintXProc,
  		  NULL);
        }
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
        // Else, if using the PostScript drawing model...
        else if (model == POSTSCRIPT) {
--- 1119,1125 ----
  		  CANVAS_REPAINT_PROC, &ComponentDisplay::repaintXProc,
  		  NULL);
        }
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
        // Else, if using the PostScript drawing model...
        else if (model == POSTSCRIPT) {
***************
*** 1201,1207 ****
    if (found) component.show (FALSE);
  }
   
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  /*F+ void ComponentDisplay::storePSDrawImageProc (char    *name,
  *                                                 Image   &image,
--- 1201,1207 ----
    if (found) component.show (FALSE);
  }
   
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  /*F+ void ComponentDisplay::storePSDrawImageProc (char    *name,
  *                                                 Image   &image,
***************
*** 1593,1599 ****
  			    WIN_BIT_GRAVITY,	       bitGravity,
  			    NULL);
    }
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
    // Else, if the drawing model is PostScript...
    else if (model == POSTSCRIPT) {
--- 1593,1599 ----
  			    WIN_BIT_GRAVITY,	       bitGravity,
  			    NULL);
    }
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
    // Else, if the drawing model is PostScript...
    else if (model == POSTSCRIPT) {
***************
*** 1677,1683 ****
    canvasXDisplay = (Display *) xv_get (paintWindow, XV_DISPLAY);
    canvasXWindow = (Window) xv_get (paintWindow, XV_XID);
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
    // If the drawing model is PostScript...
    if (model == POSTSCRIPT)
--- 1677,1683 ----
    canvasXDisplay = (Display *) xv_get (paintWindow, XV_DISPLAY);
    canvasXWindow = (Window) xv_get (paintWindow, XV_XID);
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
    // If the drawing model is PostScript...
    if (model == POSTSCRIPT)
***************
*** 1945,1951 ****
  	        CANVAS_REPAINT_PROC, &ComponentDisplay::repaintXProc,
  	        NULL);
      }
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      // Else, if using the PostScript drawing model...
      else if (model == POSTSCRIPT) {
--- 1945,1951 ----
  	        CANVAS_REPAINT_PROC, &ComponentDisplay::repaintXProc,
  	        NULL);
      }
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      // Else, if using the PostScript drawing model...
      else if (model == POSTSCRIPT) {
diff -c -r --new-file UICV3orig//UIT/src/UIC/Makefile.linux UICV3//UIT/src/UIC/Makefile.linux
*** UICV3orig//UIT/src/UIC/Makefile.linux
--- UICV3//UIT/src/UIC/Makefile.linux	Sun Aug  8 11:36:49 1993
***************
*** 0 ****
--- 1,224 ----
+ #
+ #    SCCS ID: @(#)Makefile.g++ 1.3 93/06/01 SMI
+ #
+ #    Copyright (c) 1990-1992 Sun Microsystems, Inc.  All rights reserved.
+ #    See LEGAL_NOTICE file for terms of the license.
+ #
+ #    Makefile modified by lmfken for Linux.
+ #
+ 
+ LIB     = $(UITHOME)/lib
+ INCLUDE = $(UITHOME)/include/c++
+ GENINC  = $(GENERICHOME)/include
+ 
+ CCFLAGS = -c -O -D$(SLINGSHOT) -I$(INCLUDE) -I$(GENERICHOME)/include \
+ 	-I$(OPENWINHOME)/include -I$(SSPKGHOME)/include
+ 
+ COMPILE.cc = g++ $(CCFLAGS)
+ LINK.cc = g++ $(CCFLAGS)
+ 
+ OBJECTS = UIObject.o UIWindow.o BaseWindow.o PopUpWindow.o \
+ 	   ComponentDisplay.o UIDisplay.o UIComponent.o TextItem.o \
+ 	   UIInputTextItem.o AlphanumericInput.o NumericInput.o Button.o \
+ 	   Gauge.o Slider.o Scroller.o TextDisplay.o TermDisplay.o UIChoice.o \
+ 	   ToggleChoice.o StackChoice.o CheckBoxChoice.o ListChoice.o \
+ 	   ColorMap.o Notifier.o Image.o CharacterFont.o \
+   	   UIMenu.o CommandMenu.o ChoiceMenu.o NoticePrompt.o \
+ 	   HotRegion.o ComponentGroup.o MultilineInput.o \
+ 	   DropTarget.o InputEvent.o TTYDisplay.o FileChooser.o
+ 
+ SSOBJECTS = \
+ 	CanvasShellDisplay.o CSComponent.o \
+ 	CSArrayTile.o \
+ 	CSBag.o \
+ 	CSBox.o \
+ 	CSCenterBox.o \
+ 	CSClock.o \
+ 	CSDrawArea.o \
+ 	CSDrawIcon.o \
+ 	CSDrawImage.o \
+ 	CSDrawLine.o \
+ 	CSDrawRect.o \
+ 	CSDrawText.o \
+ 	CSGrip.o \
+ 	CSTempGrip.o \
+ 	CSTree.o
+ 
+ all:	$(LIB)/libUIC.a $(LIB)/libUICSS.a
+ 
+ $(LIB)/libUIC.a: $(OBJECTS)
+ 	cp $(GENERICHOME)/lib/libGeneric.a $(LIB)/libUIC.a
+ 	ar cr $(LIB)/libUIC.a $(OBJECTS)
+ 	ranlib $(LIB)/libUIC.a
+ 	/bin/cp $(GENINC)/Generic.h $(INCLUDE)
+ 	/bin/cp $(GENINC)/GenericList.h $(INCLUDE)
+ 	/bin/cp $(GENINC)/GenericHash.h $(INCLUDE)
+ 	/bin/cp $(GENINC)/Boolean.h $(INCLUDE)
+ 
+ $(LIB)/libUICSS.a: $(SSOBJECTS)
+ 	ar cr $(LIB)/libUICSS.a $(SSOBJECTS)
+ 	ranlib $(LIB)/libUICSS.a
+ 
+ 
+ ColorMap.o:	ColorMap.cc $(INCLUDE)/ColorMap.h \
+ 		$(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h $(GENINC)/GenericList.h
+ 
+ Notifier.o:	Notifier.cc $(INCLUDE)/Notifier.h $(GENINC)/Generic.h \
+ 		$(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/BaseWindow.h 
+ 
+ Image.o:	Image.cc $(INCLUDE)/Image.h $(GENINC)/Generic.h \
+ 		$(INCLUDE)/UIObject.h
+ 
+ CharacterFont.o: CharacterFont.cc $(INCLUDE)/CharacterFont.h \
+ 		$(INCLUDE)/UIObject.h \
+ 		$(GENINC)/Generic.h
+ 
+ UIMenu.o:	UIMenu.cc $(INCLUDE)/UIMenu.h $(GENINC)/Generic.h \
+ 		$(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/ColorMap.h
+ 
+ CommandMenu.o:	CommandMenu.cc $(INCLUDE)/CommandMenu.h $(GENINC)/Generic.h \
+ 		$(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/UIMenu.h
+ 
+ ChoiceMenu.o:	ChoiceMenu.cc $(INCLUDE)/ChoiceMenu.h $(GENINC)/Generic.h \
+ 		$(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/UIMenu.h
+ 
+ UIObject.o:	UIObject.cc $(INCLUDE)/UIObject.h $(INCLUDE)/UICTypes.h \
+ 		$(GENINC)/Generic.h $(GENINC)/GenericHash.h \
+ 		$(INCLUDE)/ColorMap.h
+ 
+ UIWindow.o:	UIWindow.cc $(INCLUDE)/UIWindow.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h $(GENINC)/GenericList.h
+ 
+ UIComponent.o:	UIComponent.cc $(INCLUDE)/UIComponent.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/Image.h
+ 
+ TextItem.o:	TextItem.cc $(INCLUDE)/TextItem.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIComponent.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/Image.h
+ 
+ Button.o:	Button.cc $(INCLUDE)/Button.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIComponent.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/Image.h $(INCLUDE)/UIMenu.h
+ 
+ Gauge.o:	Gauge.cc $(INCLUDE)/Gauge.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIComponent.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/Image.h
+ 
+ UIInputTextItem.o: UIInputTextItem.cc $(INCLUDE)/UIInputTextItem.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/Image.h
+ 
+ AlphanumericInput.o: AlphanumericInput.cc $(INCLUDE)/UIInputTextItem.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/AlphanumericInput.h \
+ 		$(INCLUDE)/Image.h
+ 
+ Slider.o:	Slider.cc $(INCLUDE)/Slider.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/Gauge.h $(INCLUDE)/Image.h
+ 
+ NumericInput.o: NumericInput.cc $(INCLUDE)/UIInputTextItem.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/NumericInput.h \
+ 		$(INCLUDE)/Image.h
+ 
+ BaseWindow.o:	BaseWindow.cc $(INCLUDE)/BaseWindow.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIWindow.h $(INCLUDE)/UICTypes.h \
+  		$(GENINC)/GenericList.h $(INCLUDE)/Image.h
+ 
+ PopUpWindow.o:	PopUpWindow.cc $(INCLUDE)/PopUpWindow.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIWindow.h $(INCLUDE)/UICTypes.h \
+  		$(GENINC)/GenericList.h $(INCLUDE)/ColorMap.h
+ 
+ UIDisplay.o:	UIDisplay.cc $(INCLUDE)/UIDisplay.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UICTypes.h
+ 
+ ComponentDisplay.o:	ComponentDisplay.cc $(INCLUDE)/ComponentDisplay.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIDisplay.h \
+ 		$(INCLUDE)/UICTypes.h $(GENINC)/GenericList.h \
+ 		$(INCLUDE)/UIMenu.h
+ 
+ TextDisplay.o:	TextDisplay.cc $(INCLUDE)/TextDisplay.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIDisplay.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/ColorMap.h
+ 
+ TermDisplay.o:	TermDisplay.cc $(INCLUDE)/TermDisplay.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIDisplay.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/ColorMap.h
+ 
+ UIChoice.o:	UIChoice.cc $(INCLUDE)/UIChoice.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(GENINC)/GenericList.h \
+ 		$(INCLUDE)/Image.h $(INCLUDE)/ColorMap.h
+  
+ StackChoice.o:	StackChoice.cc $(INCLUDE)/UIChoice.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/StackChoice.h \
+ 		$(GENINC)/GenericList.h $(INCLUDE)/Image.h
+ 
+ ToggleChoice.o:	ToggleChoice.cc $(INCLUDE)/UIChoice.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/ToggleChoice.h \
+ 		$(GENINC)/GenericList.h $(INCLUDE)/Image.h
+ 
+ ListChoice.o:	ListChoice.cc $(INCLUDE)/UIChoice.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/ListChoice.h \
+ 		$(GENINC)/GenericList.h $(INCLUDE)/Image.h \
+ 	 	$(INCLUDE)/UIMenu.h
+ 
+ CheckBoxChoice.o:	CheckBoxChoice.cc $(INCLUDE)/UIChoice.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UIComponent.h \
+ 		$(INCLUDE)/UICTypes.h $(INCLUDE)/CheckBoxChoice.h \
+ 		$(GENINC)/GenericList.h $(INCLUDE)/Image.h
+ 
+ Scroller.o:	Scroller.cc $(INCLUDE)/Scroller.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIComponent.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/ColorMap.h
+ 
+ NoticePrompt.o:	NoticePrompt.cc $(INCLUDE)/NoticePrompt.h \
+ 		$(INCLUDE)/UIObject.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/ColorMap.h
+ 
+ ComponentGroup.o: ComponentGroup.cc $(INCLUDE)/ComponentGroup.h \
+                 $(INCLUDE)/UIComponent.h $(INCLUDE)/ComponentDisplay.h \
+                 $(INCLUDE)/UIObject.h $(INCLUDE)/UICTypes.h
+ 
+ MultilineInput.o: MultilineInput.cc $(INCLUDE)/MultilineInput.h \
+                 $(INCLUDE)/UIComponent.h $(INCLUDE)/UIInputTextItem.h \
+                 $(INCLUDE)/UIObject.h $(INCLUDE)/UICTypes.h
+ 
+ DropTarget.o:   DropTarget.cc $(INCLUDE)/DropTarget.h $(INCLUDE)/UIObject.h \
+                 $(INCLUDE)/UIComponent.h $(INCLUDE)/UICTypes.h \
+                 $(INCLUDE)/ColorMap.h
+ 
+ TTYDisplay.o:   TTYDisplay.cc $(INCLUDE)/TTYDisplay.h $(INCLUDE)/UIObject.h \
+ 		$(INCLUDE)/UIDisplay.h $(INCLUDE)/UICTypes.h \
+ 		$(INCLUDE)/ColorMap.h
+ 
+ FileChooser.o:  FileChooser.cc $(INCLUDE)/FileChooser.h $(INCLUDE)/UIObject.h \
+                 $(INCLUDE)/UIWindow.h $(INCLUDE)/PopUpWindow.h \
+                 $(INCLUDE)/UICTypes.h $(INCLUDE)/ColorMap.h
+  
+ delete:
+ 	/bin/rm -f $(OBJECTS)
+ 	/bin/rm -f $(SSOBJECTS)
+ 	/bin/rm -f $(TARGET)
+ 	/bin/rm -f $(INCLUDE)/Generic.h
+ 	/bin/rm -f $(INCLUDE)/GenericList.h
+ 	/bin/rm -f $(INCLUDE)/GenericHash.h
+ 	/bin/rm -f $(INCLUDE)/Boolean.h
+ 
+ clean:
+ 	/bin/rm -f $(OBJECTS)
+ 	/bin/rm -f $(SSOBJECTS)
+ 	/bin/rm -f $(INCLUDE)/Generic.h
+ 	/bin/rm -f $(INCLUDE)/GenericList.h
+ 	/bin/rm -f $(INCLUDE)/GenericHash.h
+ 	/bin/rm -f $(INCLUDE)/Boolean.h
+ 
diff -c -r --new-file UICV3orig//UIT/src/UIC/UIObject.cc UICV3//UIT/src/UIC/UIObject.cc
*** UICV3orig//UIT/src/UIC/UIObject.cc	Fri Jun 18 04:02:35 1993
--- UICV3//UIT/src/UIC/UIObject.cc	Sat Jun 26 15:39:17 1993
***************
*** 26,31 ****
--- 26,34 ----
  #ifndef __GNUC__
  #include <sysent.h>
  #endif
+ #ifdef __linux
+ #include <unistd.h>      /* gethostname() */
+ #endif
  
  #include "UIObject.h"
  #include "ColorMap.h"
***************
*** 33,39 ****
  #include "HotRegion.h"
  #include "Image.h"
  
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #include <wire/wire.h>
  #endif
--- 36,42 ----
  #include "HotRegion.h"
  #include "Image.h"
  
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
  #include <wire/wire.h>
  #endif
***************
*** 416,422 ****
  	       xv_x_error_proc, x_error_proc,
  	       NULL);
      
! #ifndef NO_XVPS
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      // If the PostScript connection attribute is specified...
      if (open_ps_connection) {
--- 419,425 ----
  	       xv_x_error_proc, x_error_proc,
  	       NULL);
      
! #if !defined(NO_XVPS) && !defined(__linux)
  #if !XV_VERSION_NUMBER || (XV_VERSION_NUMBER < 3200)
      // If the PostScript connection attribute is specified...
      if (open_ps_connection) {
diff -c -r --new-file UICV3orig//envvars UICV3//envvars
*** UICV3orig//envvars
--- UICV3//envvars	Fri Aug  6 20:31:46 1993
***************
*** 0 ****
--- 1,5 ----
+ UITHOME=`pwd`/UIT; export UITHOME
+ GENERICHOME=`pwd`/Generic; export GENERICHOME
+ SLINGSHOT=SLINGSHOT; export SLINGSHOT
+ OPENWINHOME=/usr/openwin; export OPENWINHOME
+ SSPKGHOME=/usr/openwin; export SSPKGHOME
