diff -c -r xvorig/xview3.2/config//Imakefile xview3.2/config//Imakefile
*** xvorig/xview3.2/config//Imakefile	Tue Jun 29 08:13:19 1993
--- xview3.2/config//Imakefile	Sun Aug  8 12:05:27 1993
***************
*** 8,14 ****
  /**/# config directory
  /**/# @(#)Imakefile	1.9 28 Jun 1993 SMI
  
! XVFILES = XView.cf XView.lib XView.obj XView.rules XView.tmpl library.tmpl
  CFFILES = $(XVFILES) $(HEADERS)
  
  InstallMultipleDestFlags(install,$(CFFILES),$(CONFIGDIR),$(INSTDATFLAGS))
--- 8,15 ----
  /**/# config directory
  /**/# @(#)Imakefile	1.9 28 Jun 1993 SMI
  
! #XVFILES = XView.cf XView.lib XView.obj XView.rules XView.tmpl library.tmpl
! XVFILES = XView.cf XView.lib XView.obj XView.rules XView.tmpl
  CFFILES = $(XVFILES) $(HEADERS)
  
  InstallMultipleDestFlags(install,$(CFFILES),$(CONFIGDIR),$(INSTDATFLAGS))
Only in xview3.2/config/: Makefile.bak
diff -c -r xvorig/xview3.2/config//XView.cf xview3.2/config//XView.cf
*** xvorig/xview3.2/config//XView.cf	Tue Jun 29 08:13:20 1993
--- xview3.2/config//XView.cf	Tue Aug  3 22:27:55 1993
***************
*** 84,90 ****
   *	set the following value to YES for PostScript files. 
   */
  #ifndef InstallInOpenwinHome
! #define InstallInOpenwinHome NO
  #endif
  /*
   *	Where to install the XView distribution.
--- 84,90 ----
   *	set the following value to YES for PostScript files. 
   */
  #ifndef InstallInOpenwinHome
! #define InstallInOpenwinHome YES
  #endif
  /*
   *	Where to install the XView distribution.
***************
*** 101,107 ****
   *	the appropriate value.
   */
  #ifndef XVDestDir
! #define XVDestDir /home/xview
  #endif
  /*
   *	Command to add object modules to the library.
--- 101,107 ----
   *	the appropriate value.
   */
  #ifndef XVDestDir
! #define XVDestDir /usr/openwin
  #endif
  /*
   *	Command to add object modules to the library.
***************
*** 139,145 ****
   *	installed in $(DESTDIR)/lib/X11/config (aka $(CONFIGDIR)).
   */ 
  #ifndef OnlyNeedsNewXViewConfigInstalled 
! #define OnlyNeedsNewXViewConfigInstalled NO 
  #endif 
  /*
   *	If you would like the XView man pages to be installed in 
--- 139,145 ----
   *	installed in $(DESTDIR)/lib/X11/config (aka $(CONFIGDIR)).
   */ 
  #ifndef OnlyNeedsNewXViewConfigInstalled 
! #define OnlyNeedsNewXViewConfigInstalled YES 
  #endif 
  /*
   *	If you would like the XView man pages to be installed in 
***************
*** 194,200 ****
  # if PreR4Server
  # define LibXViewDefines -DPRE_R4_ICCCM
  # else
! # define LibXViewDefines -DOW_I18N -DFULL_R5 -I$(OPENWINHOME)/include
  # endif
  #endif
  /*
--- 194,201 ----
  # if PreR4Server
  # define LibXViewDefines -DPRE_R4_ICCCM
  # else
! /* # define LibXViewDefines -OW_I18N -DFULL_R5 -I$(OPENWINHOME)/include */
! # define LibXViewDefines -I$(OPENWINHOME)/include
  # endif
  #endif
  /*
***************
*** 202,208 ****
   *	set WantToCompileSharedLibs to YES below.
   */
  #ifndef WantToCompileSharedLibs
! #define WantToCompileSharedLibs YES
  #endif
  /*
   *	To compile a shared library, set HasSharedLibraries to YES in
--- 203,209 ----
   *	set WantToCompileSharedLibs to YES below.
   */
  #ifndef WantToCompileSharedLibs
! #define WantToCompileSharedLibs NO
  #endif
  /*
   *	To compile a shared library, set HasSharedLibraries to YES in
***************
*** 258,264 ****
   *	following variable to YES.  
   */
  #ifndef InstallAllImages
! #define InstallAllImages NO
  #endif
  /*
   *	If you do not already have the textsw extras menu installed in
--- 259,265 ----
   *	following variable to YES.  
   */
  #ifndef InstallAllImages
! #define InstallAllImages YES
  #endif
  /*
   *	If you do not already have the textsw extras menu installed in
***************
*** 304,310 ****
   *	malloc() and read() will be used instead.
   */
  #ifndef OsHasMmap
! #define OsHasMmap YES
  #endif
  /*
   *	The implementation of internationalization support in XView depends
--- 305,311 ----
   *	malloc() and read() will be used instead.
   */
  #ifndef OsHasMmap
! #define OsHasMmap NO
  #endif
  /*
   *	The implementation of internationalization support in XView depends
***************
*** 340,346 ****
  
   */
  #ifndef XvI18nLevel
! #define XvI18nLevel 4
  #endif
  
  /*
--- 341,347 ----
  
   */
  #ifndef XvI18nLevel
! #define XvI18nLevel 3
  #endif
  
  /*
***************
*** 349,355 ****
   * NO, as these routines are already part of the OS.
   */
  #ifndef XvNeedGettext
! #define XvNeedGettext NO
  #endif
  
  /*
--- 350,356 ----
   * NO, as these routines are already part of the OS.
   */
  #ifndef XvNeedGettext
! #define XvNeedGettext YES
  #endif
  
  /*
***************
*** 359,365 ****
   *    	be defined as -PIC
   */
  #ifndef PositionIndependentCodeFlag
! #define PositionIndependentCodeFlag -K PIC
  #endif
  
  /*
--- 360,366 ----
   *    	be defined as -PIC
   */
  #ifndef PositionIndependentCodeFlag
! #define PositionIndependentCodeFlag -PIC
  #endif
  
  /*
***************
*** 377,383 ****
   *	acc			SunOS 4.x		-strconst
   */
  #ifndef StringConstantFlag
! #define StringConstantFlag -xstrconst
  #endif
  
  #endif /* XViewCf */
--- 378,384 ----
   *	acc			SunOS 4.x		-strconst
   */
  #ifndef StringConstantFlag
! #define StringConstantFlag
  #endif
  
  #endif /* XViewCf */
diff -c -r xvorig/xview3.2/config//XView.rules xview3.2/config//XView.rules
*** xvorig/xview3.2/config//XView.rules	Tue Jun 29 08:13:20 1993
--- xview3.2/config//XView.rules	Mon Jul  5 23:48:09 1993
***************
*** 366,374 ****
  #ifndef InstallSharedLibraryNoBuild
  #define	InstallSharedLibraryNoBuild(libname,rev,dest)				@@\
  install::  								@@\
!         $(RM) /dest/lib/**/libname.so                                   @@\
  	$(INSTALL) -c $(INSTSHAREDLIBFLAGS) lib/**/libname.so.rev dest	@@\
!         (cd dest; $(LN) lib/**/libname.so.rev lib/**/libname.so)
  
  #endif /* InstallSharedLibraryNoBuild */
  
--- 366,374 ----
  #ifndef InstallSharedLibraryNoBuild
  #define	InstallSharedLibraryNoBuild(libname,rev,dest)				@@\
  install::  								@@\
! 	$(RM) /dest/lib/**/libname.so                                   @@\
  	$(INSTALL) -c $(INSTSHAREDLIBFLAGS) lib/**/libname.so.rev dest	@@\
! 	(cd dest; $(LN) lib/**/libname.so.rev lib/**/libname.so)
  
  #endif /* InstallSharedLibraryNoBuild */
  
diff -c -r xvorig/xview3.2/config//XView.rules.xstr xview3.2/config//XView.rules.xstr
*** xvorig/xview3.2/config//XView.rules.xstr	Tue Jun 29 08:13:21 1993
--- xview3.2/config//XView.rules.xstr	Mon Jul  5 23:47:28 1993
***************
*** 385,393 ****
  #ifndef InstallSharedLibraryNoBuild
  #define	InstallSharedLibraryNoBuild(libname,rev,dest)				@@\
  install::  								@@\
!         $(RM) /dest/lib/**/libname.so                                   @@\
  	$(INSTALL) -c $(INSTSHAREDLIBFLAGS) lib/**/libname.so.rev dest	@@\
!         (cd dest; $(LN) lib/**/libname.so.rev lib/**/libname.so)
  
  #endif /* InstallSharedLibraryNoBuild */
  
--- 385,393 ----
  #ifndef InstallSharedLibraryNoBuild
  #define	InstallSharedLibraryNoBuild(libname,rev,dest)				@@\
  install::  								@@\
! 	$(RM) /dest/lib/**/libname.so                                   @@\
  	$(INSTALL) -c $(INSTSHAREDLIBFLAGS) lib/**/libname.so.rev dest	@@\
! 	(cd dest; $(LN) lib/**/libname.so.rev lib/**/libname.so)
  
  #endif /* InstallSharedLibraryNoBuild */
  
diff -c -r xvorig/xview3.2/config//XView.tmpl xview3.2/config//XView.tmpl
*** xvorig/xview3.2/config//XView.tmpl	Tue Jun 29 08:13:21 1993
--- xview3.2/config//XView.tmpl	Mon Jul  5 23:33:52 1993
***************
*** 96,129 ****
   *      Work around small conflict with Project.tmpl file
   */
  #if UseInstalled
!               MKDIRHIER = BourneShell BinDir/mkdirhier
  #endif
  
  #if OsHasLocale
!          LOCALE_DEFINES = -DOS_HAS_LOCALE
  #endif
  
  #if OsHasMmap
!            MMAP_DEFINES = -DOS_HAS_MMAP
  #endif
  
  #if XvUseAllocFuncs
!            XV_ALLOC_DEFINES = -DXV_USE_ALLOC_FUNCS
  #endif
  
  #if XvNoStrdup
!            XV_STRDUP_DEFINES = -DXV_NO_STRDUP
  #endif
  
  #if SharedLibXView 
!            XV_CFLAGS.optimized = 
!            XV_CFLAGS.shared = -DDYNAMICLIB
  #else
!            XV_CFLAGS.optimized = -DOPTIMIZED
!            XV_CFLAGS.shared =
  #endif
!                 DEFINES = $(LOCALE_DEFINES) $(MMAP_DEFINES) $(XV_ALLOC_DEFINES) \
! 			  $(XV_STRDUP_DEFINES) $(XV_CFLAGS.optimized) $(XV_CFLAGS.shared)
  
  #ifdef XVDestDir
                XVDESTDIR = XVDestDir
--- 96,129 ----
   *      Work around small conflict with Project.tmpl file
   */
  #if UseInstalled
!             MKDIRHIER = BourneShell BinDir/mkdirhier
  #endif
  
  #if OsHasLocale
!             LOCALE_DEFINES = -DOS_HAS_LOCALE
  #endif
  
  #if OsHasMmap
!             MMAP_DEFINES = -DOS_HAS_MMAP
  #endif
  
  #if XvUseAllocFuncs
!             XV_ALLOC_DEFINES = -DXV_USE_ALLOC_FUNCS
  #endif
  
  #if XvNoStrdup
!             XV_STRDUP_DEFINES = -DXV_NO_STRDUP
  #endif
  
  #if SharedLibXView 
!             XV_CFLAGS.optimized = 
!             XV_CFLAGS.shared = -DDYNAMICLIB
  #else
!             XV_CFLAGS.optimized = -DOPTIMIZED
!             XV_CFLAGS.shared =
  #endif
!             DEFINES = $(LOCALE_DEFINES) $(MMAP_DEFINES) $(XV_ALLOC_DEFINES) \
! 		  $(XV_STRDUP_DEFINES) $(XV_CFLAGS.optimized) $(XV_CFLAGS.shared)
  
  #ifdef XVDestDir
                XVDESTDIR = XVDestDir
***************
*** 356,365 ****
  
  #if SystemV4
               SYSV_CLIENT_LIB = -lintl -ldl
! 	     XVCLIENTSYSDEFINES = -DSVR4 -DSYSV
  #else
               SYSV_CLIENT_LIB =
! 	     XVCLIENTSYSDEFINES = -DSUNOS41
  #endif
  
  /*
--- 356,365 ----
  
  #if SystemV4
               SYSV_CLIENT_LIB = -lintl -ldl
!              XVCLIENTSYSDEFINES = -DSVR4 -DSYSV
  #else
               SYSV_CLIENT_LIB =
!              XVCLIENTSYSDEFINES = -DSUNOS41
  #endif
  
  /*
***************
*** 371,386 ****
   */
  
  #if XvI18nLevel == 4
! 	XVCLIENTI18NDEFINES = -DOW_I18N_L4 -DOW_I18N_L3 -DOW_I18N -DFULL_R5
! 	XVCLIENTMISCLIB = -lw
! 	XVCLIENTINCLUDES = -I$(OPENWINHOME)/include
  #else
! 	XVCLIENTI18NDEFINES = -DOW_I18N_L3
! 	XVCLIENTMISCLIB = 
! 	XVCLIENTINCLUDES = 
  #endif
  
! 	XVCLIENTDEFINES = $(XVCLIENTSYSDEFINES) $(XVCLIENTI18NDEFINES)
  
  /*
   *	Define libraries for XView clients
--- 371,386 ----
   */
  
  #if XvI18nLevel == 4
!              XVCLIENTI18NDEFINES = -DOW_I18N_L4 -DOW_I18N_L3 -DOW_I18N -DFULL_R5
!              XVCLIENTMISCLIB = -lw
!              XVCLIENTINCLUDES = -I$(OPENWINHOME)/include
  #else
!              XVCLIENTI18NDEFINES = -DOW_I18N_L3
!              XVCLIENTMISCLIB = 
!              XVCLIENTINCLUDES = 
  #endif
  
!              XVCLIENTDEFINES = $(XVCLIENTSYSDEFINES) $(XVCLIENTI18NDEFINES)
  
  /*
   *	Define libraries for XView clients
