diff -c -r --new-file xvorig/xview3.2/lib//libxview/base/Imakefile xview3.2/lib//libxview/base/Imakefile
*** xvorig/xview3.2/lib//libxview/base/Imakefile	Tue Jun 29 08:14:40 1993
--- xview3.2/lib//libxview/base/Imakefile	Tue Jul  6 00:31:30 1993
***************
*** 37,43 ****
  	drawable.o\
  	xv_debug.o\
  	xv_parse.o\
! 	xv_list.o\ 
  	xv_error.o
  
  SRCS=\
--- 37,43 ----
  	drawable.o\
  	xv_debug.o\
  	xv_parse.o\
! 	xv_list.o\
  	xv_error.o
  
  SRCS=\
diff -c -r --new-file xvorig/xview3.2/lib//libxview/base/base.h xview3.2/lib//libxview/base/base.h
*** xvorig/xview3.2/lib//libxview/base/base.h	Tue Jun 29 08:14:42 1993
--- xview3.2/lib//libxview/base/base.h	Tue Oct  5 21:55:12 1993
***************
*** 53,62 ****
   * the macro SUNOS41. This has to be fixed later to accomodate non-SUNOS41 and
   * non-SVR4 systems.
   */
! #ifndef SUNOS41
  #define XV_OS_SVR4
  #define XV_USE_TTCOMPAT
  #define SYSV_WAIT 
  #define SYSV_UCONTEXT 
  #define XV_USE_XVFCNTL 
  #endif
--- 53,67 ----
   * the macro SUNOS41. This has to be fixed later to accomodate non-SUNOS41 and
   * non-SVR4 systems.
   */
! #if !defined(SUNOS41) && !defined(__linux)
  #define XV_OS_SVR4
  #define XV_USE_TTCOMPAT
  #define SYSV_WAIT 
+ #define SYSV_UCONTEXT 
+ #define XV_USE_XVFCNTL 
+ #elif defined(__linux)
+ #define XV_OS_SVR4
+ #undef XV_USE_TTCOMPAT
  #define SYSV_UCONTEXT 
  #define XV_USE_XVFCNTL 
  #endif
diff -c -r --new-file xvorig/xview3.2/lib//libxview/base/xv_debug.h xview3.2/lib//libxview/base/xv_debug.h
*** xvorig/xview3.2/lib//libxview/base/xv_debug.h	Tue Jun 29 08:14:44 1993
--- xview3.2/lib//libxview/base/xv_debug.h	Sun Jul 11 19:48:17 1993
***************
*** 10,16 ****
  #define _xv_debug_h_already_included
  
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 10,16 ----
  #define _xv_debug_h_already_included
  
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/base/xv_error.c xview3.2/lib//libxview/base/xv_error.c
*** xvorig/xview3.2/lib//libxview/base/xv_error.c	Tue Jun 29 08:14:42 1993
--- xview3.2/lib//libxview/base/xv_error.c	Tue Aug  3 05:37:45 1993
***************
*** 21,27 ****
--- 21,32 ----
  extern char    *sys_errlist[];
  extern int      errno;
  
+ #ifndef __linux
+ /* Global already defined in xv_init.c */
  Xv_private_data char *xv_app_name;
+ #else
+ extern Xv_private_data char *xv_app_name;
+ #endif
  Xv_private int (*xv_error_proc) ();
  
  /*
diff -c -r --new-file xvorig/xview3.2/lib//libxview/dnd/site.c xview3.2/lib//libxview/dnd/site.c
*** xvorig/xview3.2/lib//libxview/dnd/site.c	Tue Jun 29 08:15:58 1993
--- xview3.2/lib//libxview/dnd/site.c	Sat Jul 10 23:00:28 1993
***************
*** 241,247 ****
  	      return(XV_ERROR);
  
  	  rect = xv_alloc(Rect);
! #ifdef SVR4
            /* This will probably not work right, but it compiles. */
            /* (rectNode->rect) is of the wrong type. */
            memmove(rect, &(rectNode->rect),  sizeof(Rect));
--- 241,247 ----
  	      return(XV_ERROR);
  
  	  rect = xv_alloc(Rect);
! #if defined(SVR4) || defined(__linux)
            /* This will probably not work right, but it compiles. */
            /* (rectNode->rect) is of the wrong type. */
            memmove(rect, &(rectNode->rect),  sizeof(Rect));
diff -c -r --new-file xvorig/xview3.2/lib//libxview/file_chooser/Imakefile xview3.2/lib//libxview/file_chooser/Imakefile
*** xvorig/xview3.2/lib//libxview/file_chooser/Imakefile	Tue Jun 29 08:17:58 1993
--- xview3.2/lib//libxview/file_chooser/Imakefile	Tue Jul  6 00:32:45 1993
***************
*** 40,47 ****
          hist_list.o\
          hist_menu.o\
  	fchsr_data.o\
!         flist_data.o\ 
!         path_data.o\ 
          hist_data.o
  
  
--- 40,47 ----
          hist_list.o\
          hist_menu.o\
  	fchsr_data.o\
!         flist_data.o\
!         path_data.o\
          hist_data.o
  
  
***************
*** 54,61 ****
  	hist_list.c\
  	hist_menu.c\
  	fchsr_data.c\
!         flist_data.c\ 
!         path_data.c\ 
          hist_data.c
  
  LIBTARGET = xview
--- 54,61 ----
  	hist_list.c\
  	hist_menu.c\
  	fchsr_data.c\
!         flist_data.c\
!         path_data.c\
          hist_data.c
  
  LIBTARGET = xview
diff -c -r --new-file xvorig/xview3.2/lib//libxview/file_chooser/fc_layout.c xview3.2/lib//libxview/file_chooser/fc_layout.c
*** xvorig/xview3.2/lib//libxview/file_chooser/fc_layout.c	Tue Jun 29 08:17:55 1993
--- xview3.2/lib//libxview/file_chooser/fc_layout.c	Thu Jul 22 21:05:03 1993
***************
*** 464,471 ****
  
  	new_height 
  	    = (* private->exten_func)( FC_PUBLIC(private),
! 				      private->rect,
! 				      exten_rect,
  				      COLS(2),
  				      private->rect.r_width - COLS(2),
  				      max_height
--- 464,471 ----
  
  	new_height 
  	    = (* private->exten_func)( FC_PUBLIC(private),
! 				      &private->rect,
! 				      &exten_rect,
  				      COLS(2),
  				      private->rect.r_width - COLS(2),
  				      max_height
diff -c -r --new-file xvorig/xview3.2/lib//libxview/file_chooser/file_list.c xview3.2/lib//libxview/file_chooser/file_list.c
*** xvorig/xview3.2/lib//libxview/file_chooser/file_list.c	Tue Jun 29 08:17:56 1993
--- xview3.2/lib//libxview/file_chooser/file_list.c	Sun Jul 11 01:12:20 1993
***************
*** 521,526 ****
--- 521,530 ----
      if (status == DESTROY_CLEANUP) {
  	xv_free_ref( private->directory );
  	xv_free_ref( private->regex_pattern );
+ #ifdef __linux
+ 	if (private->regex_compile != NULL && private->regex_compile->allocated)
+ 		xv_free_ref( private->regex_compile->buffer);
+ #endif
  	xv_free_ref( private->regex_compile );
  	xv_free_ref( private->dotdot_string );
  	if ( private->dir_ptr )
***************
*** 1138,1143 ****
--- 1142,1148 ----
  
  /****************************************************************************/
  
+ #ifndef __linux
  /*
   * Front end to regexp(3).
   *
***************
*** 1208,1213 ****
--- 1213,1250 ----
      return step(s, private->regex_compile);
  }
  
+ #else /* __linux */
+ 
+ /* Linux does not have regexp.h or compile()/step(). Use regex.h and
+  * re_compile_pattern()/re_match() instead. */
+ 
+ static void
+ flist_compile_regex( private )
+      File_list_private *private;
+ {
+ 
+     if (private->regex_compile == NULL) {
+         private->regex_compile = xv_alloc_n(regex_t, 1);
+         private->regex_compile->translate = NULL;
+     }
+     if (private->regex_compile->allocated == 0) {
+         private->regex_compile->buffer = xv_alloc_n(char, MAXPATHLEN + 1);
+         private->regex_compile->allocated = MAXPATHLEN + 1;
+     }
+     re_compile_pattern(private->regex_pattern, strlen(private->regex_pattern),
+ 			private->regex_compile);
+ } 
+ 
+ static int
+ flist_match_regex( s, private )
+      char *s;
+      File_list_private *private;
+ {
+     if (private->regex_compile == NULL || private->regex_compile->allocated == 0)
+         return 0;
+     return (re_match(private->regex_compile, s, strlen(s), 0, NULL) != -1);
+ }
+ #endif /* __linux */
  /****************************************************************************/
  
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/file_chooser/flist_impl.h xview3.2/lib//libxview/file_chooser/flist_impl.h
*** xvorig/xview3.2/lib//libxview/file_chooser/flist_impl.h	Tue Jun 29 08:17:56 1993
--- xview3.2/lib//libxview/file_chooser/flist_impl.h	Sun Jul 11 00:16:49 1993
***************
*** 9,14 ****
--- 9,17 ----
  
  #include <sys/param.h>
  #include <dirent.h>
+ #ifdef __linux
+ #include <regex.h>
+ #endif
  #include <xview_private/xv_path_util.h>
  #include <xview/file_list.h>
  #include <xview_private/i18n_impl.h>
***************
*** 22,28 ****
--- 25,35 ----
      DIR *		dir_ptr;		/* cached directory pointer */
      char *		previous_dir;		/* last directory displayed */
      char *		regex_pattern; 		/* regular expression */
+ #ifndef __linux
      char *		regex_compile;		/* compiled form of regex */
+ #else
+     regex_t *           regex_compile;          /* compiled form of regex */
+ #endif
      Server_image	match_glyph;		/* matched regex glyph */
      Server_image	match_glyph_mask;	/* matched regex glyph mask */
      unsigned short	filter_mask;		/* mask for filter_func */
diff -c -r --new-file xvorig/xview3.2/lib//libxview/frame/fm_impl.h xview3.2/lib//libxview/frame/fm_impl.h
*** xvorig/xview3.2/lib//libxview/frame/fm_impl.h	Tue Jun 29 08:16:15 1993
--- xview3.2/lib//libxview/frame/fm_impl.h	Sun Jul 11 19:50:13 1993
***************
*** 18,24 ****
  
  /* standard includes */
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 18,24 ----
  
  /* standard includes */
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/frame/fm_props.h xview3.2/lib//libxview/frame/fm_props.h
*** xvorig/xview3.2/lib//libxview/frame/fm_props.h	Tue Jun 29 08:16:14 1993
--- xview3.2/lib//libxview/frame/fm_props.h	Sun Jul 11 19:50:36 1993
***************
*** 14,20 ****
  
  /* standard includes */
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 14,20 ----
  
  /* standard includes */
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/frame/fm_rescale.c xview3.2/lib//libxview/frame/fm_rescale.c
*** xvorig/xview3.2/lib//libxview/frame/fm_rescale.c	Tue Jun 29 08:16:12 1993
--- xview3.2/lib//libxview/frame/fm_rescale.c	Tue Jul 20 20:36:04 1993
***************
*** 43,49 ****
       */
      window_default_event_func(frame_public, (Event *) 0, scale, (Notify_event_type) 0);
      window_calculate_new_size(frame_public, frame_public, &frame_height, &frame_width);
!     xv_set(frame_public, WIN_RECT, 0);
  
  
      /*
--- 43,51 ----
       */
      window_default_event_func(frame_public, (Event *) 0, scale, (Notify_event_type) 0);
      window_calculate_new_size(frame_public, frame_public, &frame_height, &frame_width);
! #ifndef __linux
!     xv_set(frame_public, WIN_RECT, 0); /* This looks like a XView bug to me */
! #endif
  
  
      /*
diff -c -r --new-file xvorig/xview3.2/lib//libxview/frame/frame_base.h xview3.2/lib//libxview/frame/frame_base.h
*** xvorig/xview3.2/lib//libxview/frame/frame_base.h	Tue Jun 29 08:16:14 1993
--- xview3.2/lib//libxview/frame/frame_base.h	Sun Jul 11 19:51:12 1993
***************
*** 18,24 ****
  
  /* standard includes */
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 18,24 ----
  
  /* standard includes */
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/frame/frame_cmd.h xview3.2/lib//libxview/frame/frame_cmd.h
*** xvorig/xview3.2/lib//libxview/frame/frame_cmd.h	Tue Jun 29 08:16:14 1993
--- xview3.2/lib//libxview/frame/frame_cmd.h	Sun Jul 11 19:51:38 1993
***************
*** 18,24 ****
  
  /* standard includes */
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 18,24 ----
  
  /* standard includes */
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/frame/frame_help.h xview3.2/lib//libxview/frame/frame_help.h
*** xvorig/xview3.2/lib//libxview/frame/frame_help.h	Tue Jun 29 08:16:10 1993
--- xview3.2/lib//libxview/frame/frame_help.h	Sun Jul 11 19:52:04 1993
***************
*** 18,24 ****
  
  /* standard includes */
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 18,24 ----
  
  /* standard includes */
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/help/help_file.c xview3.2/lib//libxview/help/help_file.c
*** xvorig/xview3.2/lib//libxview/help/help_file.c	Tue Jun 29 08:16:17 1993
--- xview3.2/lib//libxview/help/help_file.c	Sat Sep  4 16:58:37 1993
***************
*** 18,24 ****
--- 18,28 ----
  
  #include <xview_private/i18n_impl.h>
  
+ #ifndef __linux
  #define DEFAULT_HELP_DIRECTORY "/usr/lib/help"
+ #else
+ #define DEFAULT_HELP_DIRECTORY "/usr/openwin/lib/help"
+ #endif
  #define MAX_MORE_HELP_CMD 128
  
  Xv_private char *xv_strtok();
***************
*** 82,89 ****
--- 86,98 ----
     /*
      * Need to fix this to get the XV_LC_DISPLAY_LANG from server
      */
+ #ifndef __linux
      if (_xv_use_locale)
      	xv_lc_display_lang = setlocale(LC_MESSAGES, NULL); 
+ #else
+ /* The setlocale() above should be disabled if I18N not supported, a bug?? */
+     xv_lc_display_lang = "C";
+ #endif
      helpdir = xv_strtok(helppath_copy, ":");
      do {
  	/*  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/menu/Imakefile xview3.2/lib//libxview/menu/Imakefile
*** xvorig/xview3.2/lib//libxview/menu/Imakefile	Tue Jun 29 08:16:27 1993
--- xview3.2/lib//libxview/menu/Imakefile	Tue Jul  6 00:35:45 1993
***************
*** 28,34 ****
  TEXT_OBJS =\
  	omi.o\
  	om_public.o\
! 	om_render.o\ 
  	om_set.o\
  	om_get.o\
  	om_compat.o
--- 28,34 ----
  TEXT_OBJS =\
  	omi.o\
  	om_public.o\
! 	om_render.o\
  	om_set.o\
  	om_get.o\
  	om_compat.o
diff -c -r --new-file xvorig/xview3.2/lib//libxview/menu/om_render.c xview3.2/lib//libxview/menu/om_render.c
*** xvorig/xview3.2/lib//libxview/menu/om_render.c	Tue Jun 29 08:16:26 1993
--- xview3.2/lib//libxview/menu/om_render.c	Tue Jul 20 23:22:00 1993
***************
*** 470,478 ****
       * Define the menu and shadow window dimensions.  Note: shadow rect width &
       * height = menu rect width & height
       */
!     xv_set(m->window, XV_RECT, m->fs_menurect, 0);
      if (!m->group_info->three_d)
! 	xv_set(m->shadow_window, XV_RECT, shadowrect, 0);
  
      XFlush(XV_DISPLAY_FROM_WINDOW(m->window));
  
--- 470,478 ----
       * Define the menu and shadow window dimensions.  Note: shadow rect width &
       * height = menu rect width & height
       */
!     xv_set(m->window, XV_RECT, &m->fs_menurect, 0);
      if (!m->group_info->three_d)
! 	xv_set(m->shadow_window, XV_RECT, &shadowrect, 0);
  
      XFlush(XV_DISPLAY_FROM_WINDOW(m->window));
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/misc/Imakefile xview3.2/lib//libxview/misc/Imakefile
*** xvorig/xview3.2/lib//libxview/misc/Imakefile	Tue Jun 29 08:16:36 1993
--- xview3.2/lib//libxview/misc/Imakefile	Tue Jul  6 00:54:41 1993
***************
*** 57,63 ****
  	demorandom.o\
  	getlogindr.o\
  	expandname.o\
! 	expandpath.o\ 
  	bitmask.o\
  	hashfn.o\
  	db_conv.o\
--- 57,63 ----
  	demorandom.o\
  	getlogindr.o\
  	expandname.o\
! 	expandpath.o\
  	bitmask.o\
  	hashfn.o\
  	db_conv.o\
***************
*** 76,82 ****
  	demorandom.c\
  	getlogindr.c\
  	expandname.c\
! 	expandpath.c\ 
  	bitmask.c\
  	hashfn.c\
  	db_conv.c\
--- 76,82 ----
  	demorandom.c\
  	getlogindr.c\
  	expandname.c\
! 	expandpath.c\
  	bitmask.c\
  	hashfn.c\
  	db_conv.c\
***************
*** 95,101 ****
  	demorandom.o\
  	getlogindr.o\
  	expandname.o\
! 	expandpath.o\ 
  	bitmask.o\
  	hashfn.o\
  	db_conv.o\
--- 95,101 ----
  	demorandom.o\
  	getlogindr.o\
  	expandname.o\
! 	expandpath.o\
  	bitmask.o\
  	hashfn.o\
  	db_conv.o\
***************
*** 115,121 ****
  	demorandom.c\
  	getlogindr.c\
  	expandname.c\
! 	expandpath.c\ 
  	bitmask.c\
  	hashfn.c\
  	db_conv.c\
--- 115,121 ----
  	demorandom.c\
  	getlogindr.c\
  	expandname.c\
! 	expandpath.c\
  	bitmask.c\
  	hashfn.c\
  	db_conv.c\
diff -c -r --new-file xvorig/xview3.2/lib//libxview/misc/getlogindr.c xview3.2/lib//libxview/misc/getlogindr.c
*** xvorig/xview3.2/lib//libxview/misc/getlogindr.c	Tue Jun 29 08:16:30 1993
--- xview3.2/lib//libxview/misc/getlogindr.c	Sun Jul 11 02:18:17 1993
***************
*** 24,30 ****
--- 24,32 ----
  xv_getlogindir()
  {
      extern char    *getlogin(), *getenv();
+ #ifndef __linux
      extern struct passwd *getpwnam(), *getpwuid();
+ #endif
      struct passwd  *passwdent;
      char           *home, *loginname;
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/misc/gettext.h xview3.2/lib//libxview/misc/gettext.h
*** xvorig/xview3.2/lib//libxview/misc/gettext.h	Tue Jun 29 08:16:31 1993
--- xview3.2/lib//libxview/misc/gettext.h	Sun Jul 11 17:50:37 1993
***************
*** 1,7 ****
--- 1,11 ----
  /* @(#)gettext.h 50.11 93/06/28 SMI */
  
  #define DEFAULT_DOMAIN	"default"
+ #ifndef __linux
  #define DEFAULT_BINDING "/usr/lib/locale\n"
+ #else
+ #define DEFAULT_BINDING "/usr/openwin/lib/locale\n"
+ #endif
  #define COOKIE 0xFF
  #define BINDINGLISTDELIM '\n'
  
***************
*** 13,18 ****
--- 17,25 ----
  #include <errno.h>
  #ifdef OS_HAS_LOCALE
  #include <locale.h>
+ #if defined(__linux) && !defined(LC_MESSAGES) && defined(LC_RESPONSE)
+ #define LC_MESSAGES LC_RESPONSE
+ #endif
  #endif OS_HAS_LOCALE
  #include <stdio.h>
  #include <sys/types.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/misc/i18n_impl.h xview3.2/lib//libxview/misc/i18n_impl.h
*** xvorig/xview3.2/lib//libxview/misc/i18n_impl.h	Tue Jun 29 08:16:31 1993
--- xview3.2/lib//libxview/misc/i18n_impl.h	Sun Jul 11 10:53:03 1993
***************
*** 25,30 ****
--- 25,35 ----
  
  #include  <locale.h>
  
+ /* Linux: gcc 2.4.x does not have LC_MESSAGES, but it has LC_RESPONSE instead */
+ #if defined(__linux) && !defined(LC_MESSAGES) && defined(LC_RESPONSE)
+ #define LC_MESSAGES LC_RESPONSE
+ #endif
+ 
  extern char	*dgettext();
  
  #define XV_I18N_MSG(d,s)	(dgettext(d,s))
diff -c -r --new-file xvorig/xview3.2/lib//libxview/misc/portable.h xview3.2/lib//libxview/misc/portable.h
*** xvorig/xview3.2/lib//libxview/misc/portable.h	Tue Jun 29 08:16:34 1993
--- xview3.2/lib//libxview/misc/portable.h	Sun Jul 11 10:31:34 1993
***************
*** 49,55 ****
  #define const
  #endif
  
! #ifdef SVR4
  #define XV_BCOPY(a,b,c) memmove(b,a,c)
  #define XV_BZERO(a,b) memset(a,0,b)
  #define XV_INDEX(a,b) strchr(a,b)
--- 49,55 ----
  #define const
  #endif
  
! #if defined(SVR4) || defined(__linux)
  #define XV_BCOPY(a,b,c) memmove(b,a,c)
  #define XV_BZERO(a,b) memset(a,0,b)
  #define XV_INDEX(a,b) strchr(a,b)
***************
*** 66,71 ****
--- 66,75 ----
   * Defines governing tty mode and pty behavior.  (These are relevant to the
   * ttysw code.)
   */
+ #ifdef __linux
+ #define	XV_USE_TERMIOS
+ #undef	XV_USE_SVR4_PTYS
+ #else
  #ifdef	SVR4
  #define	XV_USE_TERMIOS
  #define	XV_USE_SVR4_PTYS
***************
*** 73,77 ****
--- 77,82 ----
  #undef	XV_USE_TERMIOS
  #undef	XV_USE_SVR4_PTYS
  #endif	/* SVR4 */
+ #endif  /* __linux */
  
  #endif /* xview_portable_h_DEFINED */
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notice/notice_pt.c xview3.2/lib//libxview/notice/notice_pt.c
*** xvorig/xview3.2/lib//libxview/notice/notice_pt.c	Tue Jun 29 08:16:40 1993
--- xview3.2/lib//libxview/notice/notice_pt.c	Tue Aug  3 06:00:59 1993
***************
*** 29,35 ****
--- 29,40 ----
  /*
   * Table containing valid values for OpenWindows.KeyboardCommands resource
   */
+ #ifndef __linux
  Xv_private_data Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #else
+ /* Global already defined and initialized in server/server.c */
+ extern Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #endif
  
  /*
   * --------------------------- Cursor Stuff -------------------------
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/Imakefile xview3.2/lib//libxview/notify/Imakefile
*** xvorig/xview3.2/lib//libxview/notify/Imakefile	Tue Jun 29 08:18:21 1993
--- xview3.2/lib//libxview/notify/Imakefile	Sun Jul 18 19:54:21 1993
***************
*** 48,54 ****
  	nintr_wait.o nintrdeath.o nintremove.o nintrevent.o nintrexcpt.o \
  	nintritimr.o ntfy_cond.o ntfy_ctbl.o ntfy_debug.o ntfy_dump.o \
  	ntfy_fd_op.o ntfy_list.o ntfy_node.o ntfyclient.o ntfyperror.o \
! 	ntfyprotec.o sys_fcntl.o sys_read.o sys_select.o
  	
  SRCS= \
  	ndet_auto.c ndet_death.c ndet_die.c ndet_dodis.c ndet_event.c \
--- 48,54 ----
  	nintr_wait.o nintrdeath.o nintremove.o nintrevent.o nintrexcpt.o \
  	nintritimr.o ntfy_cond.o ntfy_ctbl.o ntfy_debug.o ntfy_dump.o \
  	ntfy_fd_op.o ntfy_list.o ntfy_node.o ntfyclient.o ntfyperror.o \
! 	ntfyprotec.o sys_fcntl.o sys_read.o sys_select.o linux_select.o
  	
  SRCS= \
  	ndet_auto.c ndet_death.c ndet_die.c ndet_dodis.c ndet_event.c \
***************
*** 69,76 ****
  	nintr_wait.c nintrdeath.c nintremove.c nintrevent.c nintrexcpt.c \
  	nintritimr.c notifydata.c ntfy_cond.c ntfy_ctbl.c ntfy_debug.c \
  	ntfy_dump.c ntfy_fd_op.c ntfy_list.c ntfy_node.c ntfyclient.c \
! 	ntfyperror.c ntfyprotec.c sys_fcntl.c sys_read.c sys_select.c
! 
  
  
  LIBTARGET = xview
--- 69,76 ----
  	nintr_wait.c nintrdeath.c nintremove.c nintrevent.c nintrexcpt.c \
  	nintritimr.c notifydata.c ntfy_cond.c ntfy_ctbl.c ntfy_debug.c \
  	ntfy_dump.c ntfy_fd_op.c ntfy_list.c ntfy_node.c ntfyclient.c \
! 	ntfyperror.c ntfyprotec.c sys_fcntl.c sys_read.c sys_select.c \
! 	linux_select.c
  
  
  LIBTARGET = xview
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/linux_select.c xview3.2/lib//libxview/notify/linux_select.c
*** xvorig/xview3.2/lib//libxview/notify/linux_select.c
--- xview3.2/lib//libxview/notify/linux_select.c	Wed Sep 15 16:12:02 1993
***************
*** 0 ****
--- 1,63 ----
+ /* Function for calling the select(2) system call in linux.
+  * Linux doesn't have a syscall() function, this replaces it
+  * for select(), fcntl() and read().
+  * Kudos to Rick Sladkey (jrs@world.std.com) for suggesting
+  * this method. */
+ 
+ #include <sys/time.h>
+ 
+ /* #define __LIBRARY__ */
+ #include <linux/unistd.h>
+ /* #undef __LIBRARY__ */
+ 
+ #ifdef DEBUG
+ #include <stdio.h>
+ #endif
+ 
+ #define __NR_sys_select __NR_select
+ #define __NR_sys_fcntl __NR_fcntl
+ #define __NR_sys_read __NR_read
+ 
+ /* Create function sys_select(), which can be called instead of
+  * syscall(SYS_select, args...) */
+ 
+ _syscall1(int,sys_select, unsigned long *, buffer);
+ 
+ int linux_select(int width, fd_set *readfds, fd_set *writefds,
+                  fd_set *exceptfds, struct timeval *timeout) {
+   unsigned long selargs[5];
+   static struct timeval tout_copy;
+ 
+ #ifdef DEBUG
+   fprintf(stderr, "linux_select(%ld, %ld, %ld, %ld, %ld)...",
+           (unsigned long)width, (unsigned long)readfds,
+           (unsigned long)writefds, (unsigned long)exceptfds,
+           (unsigned long)timeout);
+ #endif
+   selargs[0] = (unsigned long)width;
+   selargs[1] = (unsigned long)readfds;
+   selargs[2] = (unsigned long)writefds;
+   selargs[3] = (unsigned long)exceptfds;
+   if (timeout != NULL) {
+     tout_copy = *timeout;
+     selargs[4] = (unsigned long)&tout_copy;
+   }
+   else
+     selargs[4] = (unsigned long)timeout;
+ #ifndef DEBUG
+   return sys_select(selargs);
+ #else
+   {
+     int res = sys_select(selargs);
+     fprintf(stderr, "returned %d\n", res);
+     return res;
+   }
+ #endif
+ }
+ 
+ /* Replacement for syscall(SYS_fcntl,...) */
+ _syscall3(int, sys_fcntl, int, fildes, int, cmd, int, arg);
+ 
+ /* Replacement for syscall(SYS_read,...) */
+ _syscall3(int, sys_read, int, fildes, char *, buf, off_t, cnt);
+ 
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/linux_select.h xview3.2/lib//libxview/notify/linux_select.h
*** xvorig/xview3.2/lib//libxview/notify/linux_select.h
--- xview3.2/lib//libxview/notify/linux_select.h	Wed Jul 14 02:14:31 1993
***************
*** 0 ****
--- 1,16 ----
+ /* Header for linux versions of system calls which are
+  * called through syscall() in XView. We don't use syscall,
+  * but the unistd.h macros in the kernel sources. */
+ 
+ #ifndef __LINUX_SELECT_H
+ #define __LINUX_SELECT_H
+ 
+ #include <sys/types.h>
+ #include <sys/time.h>
+ 
+ int linux_select(int width, fd_set *readfds, fd_set *writefds,
+                  fd_set *exceptfds, struct timeval *timeout);
+ int sys_fcntl(int fildes, int cmd, int arg);
+ int sys_read(int fildes, char * buf, off_t cnt);
+ 
+ #endif __LINUX_SELECT_H
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ndet_auto.c xview3.2/lib//libxview/notify/ndet_auto.c
*** xvorig/xview3.2/lib//libxview/notify/ndet_auto.c	Tue Jun 29 08:18:19 1993
--- xview3.2/lib//libxview/notify/ndet_auto.c	Tue Oct  5 21:52:49 1993
***************
*** 73,79 ****
--- 73,82 ----
  		/* Auto sig is async */);
      /* Sweep all conditions & send notifications (recursive enumeration) */
      switch (condition->data.signal) {
+ #ifndef __linux
+ /* In linux SIGIO == SIGURG */
        case SIGIO:
+ #endif
        case SIGURG:{
  	    NDET_ENUM_SEND  enum_send2;
  	    register int    loop_limiter = 0;
***************
*** 148,154 ****
  
  	    enum_send->wait3 = &wd;
  	    /* Look for as many children as have changed state */
! #ifndef SVR4
  	    while ((wd.pid = wait3(&wd.status, WNOHANG | WUNTRACED,
  				   &wd.rusage)) > 0)
  #else
--- 151,157 ----
  
  	    enum_send->wait3 = &wd;
  	    /* Look for as many children as have changed state */
! #if !defined(SVR4) && !defined(__linux)
  	    while ((wd.pid = wait3(&wd.status, WNOHANG | WUNTRACED,
  				   &wd.rusage)) > 0)
  #else
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ndet_fcntl.c xview3.2/lib//libxview/notify/ndet_fcntl.c
*** xvorig/xview3.2/lib//libxview/notify/ndet_fcntl.c	Tue Jun 29 08:18:03 1993
--- xview3.2/lib//libxview/notify/ndet_fcntl.c	Wed Jul 14 01:29:34 1993
***************
*** 22,29 ****
  #ifdef SVR4
  #include <sys/file.h>
  #endif SVR4
  
! extern int
  #ifdef SVR4
  xv_fcntl(fd, cmd, arg)
  #else
--- 22,33 ----
  #ifdef SVR4
  #include <sys/file.h>
  #endif SVR4
+ #ifdef __linux
+ #include <stdarg.h>
+ #endif
  
! int
! #ifndef __linux
  #ifdef SVR4
  xv_fcntl(fd, cmd, arg)
  #else
***************
*** 33,39 ****
--- 37,55 ----
  {
      fd_set          bit;
      int             res;
+ #else /* __linux */
+ /* fcntl() is declared using variable args in linux */
+ fcntl(int fd, int cmd, ...) {
+     fd_set          bit;
+     int             res;
+     va_list         args;
+     int             arg;
  
+     va_start(args, cmd);
+     arg = va_arg(args, int);
+     va_end(args);
+ #endif /* __linux */
+ 
      /* Set fd bit */
      FD_ZERO(&bit);
      FD_SET(fd, &bit);
***************
*** 58,65 ****
--- 74,86 ----
  	if (cmd == F_GETFL)
  	    arg = res;
  	NTFY_BEGIN_CRITICAL;
+ #if defined(__linux) && !defined(FNDELAY)
+ 	if (arg & O_NONBLOCK)
+ 	    FD_SET(fd, &ndet_fndelay_mask);
+ #else
  	if (arg & FNDELAY)
  	    FD_SET(fd, &ndet_fndelay_mask);
+ #endif
  #ifdef FNBIO
  	else if (arg & FNBIO)
  	    FD_SET(fd, &ndet_fndelay_mask);
***************
*** 66,79 ****
--- 87,104 ----
  #endif
  	else
  	    FD_CLR(fd, &ndet_fndelay_mask);
+ #if !defined(__linux) || defined(FASYNC)
  	if (arg & FASYNC)
  	    FD_SET(fd, &ndet_fndelay_mask);
  	else
  	    FD_CLR(fd, &ndet_fndelay_mask);
+ #endif
  	/* Make sure that are catching async related signals now */
  	if (ntfy_fd_anyset(&ndet_fasync_mask)) {
  	    ndet_enable_sig(SIGIO);
+ #ifndef __linux
  	    ndet_enable_sig(SIGURG);
+ #endif
  	}
  	/*
  	 * Setting NDET_FD_CHANGE will "fix up" signals being caught to be
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ndet_loop.c xview3.2/lib//libxview/notify/ndet_loop.c
*** xvorig/xview3.2/lib//libxview/notify/ndet_loop.c	Tue Jun 29 08:18:19 1993
--- xview3.2/lib//libxview/notify/ndet_loop.c	Sun Sep 19 01:30:19 1993
***************
*** 19,24 ****
--- 19,25 ----
  #include <xview_private/ndet.h>
  #include <xview_private/nint.h>
  #include <xview_private/ndis.h>	/* For ndis_dispatch */
+ #ifndef __linux
  #ifndef SVR4
  #include <syscall.h>
  #else SVR4
***************
*** 25,30 ****
--- 26,34 ----
  #include <sys/syscall.h>
  #include <sys/poll.h>
  #endif SVR4
+ #else /* __linux */
+ #include "linux_select.h"
+ #endif
  #include <fcntl.h>
  #include <errno.h>
  #include <signal.h>
***************
*** 55,61 ****
  
  /* NOTE! This assumes NSIG is 32. Not very portable */
  /* ndet_prev_sigvec needs to start off at all zeros */
! #ifndef SVR4
  pkg_private_data struct sigvec ndet_prev_sigvec[NSIG] = {
      {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0},
      {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0},
--- 59,65 ----
  
  /* NOTE! This assumes NSIG is 32. Not very portable */
  /* ndet_prev_sigvec needs to start off at all zeros */
! #if !defined(SVR4) && !defined(__linux)
  pkg_private_data struct sigvec ndet_prev_sigvec[NSIG] = {
      {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0},
      {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0},
***************
*** 76,88 ****
  pkg_private void ndet_signal_catcher();
  #endif
  
! #ifndef SVR4
  pkg_private_data struct sigvec ndet_sigvec = {ndet_signal_catcher, 0, 0};
  static int      ndet_signal_code;
  static struct sigcontext *ndet_signal_context;
  #else SVR4
  pkg_private_data struct sigaction ndet_sigvec =
  	{SA_RESTART, {ndet_signal_catcher}, {0}, {0,0}};
  	static int      ndet_signal_code;
  	static ucontext_t *ndet_signal_context;
  #endif SVR4
--- 80,96 ----
  pkg_private void ndet_signal_catcher();
  #endif
  
! #if !defined(SVR4) && !defined(__linux)
  pkg_private_data struct sigvec ndet_sigvec = {ndet_signal_catcher, 0, 0};
  static int      ndet_signal_code;
  static struct sigcontext *ndet_signal_context;
  #else SVR4
  pkg_private_data struct sigaction ndet_sigvec =
+ #ifndef __linux
  	{SA_RESTART, {ndet_signal_catcher}, {0}, {0,0}};
+ #else
+ 	{ndet_signal_catcher,0,SA_RESTART,NULL}; /* handler,mask,flags,restorer */
+ #endif
  	static int      ndet_signal_code;
  	static ucontext_t *ndet_signal_context;
  #endif SVR4
***************
*** 225,233 ****
--- 233,246 ----
  	     * select, not ON THE WAY into select).
  	     */
  #ifndef SVR4
+ #ifndef __linux
  	    nfds = syscall(SYS_select,
  			   FD_SETSIZE, &ibits, &obits, &ebits,
  		 (sigisempty(&ndet_sigs_received)) ? timer : &ndet_polling_tv);
+ #else
+ 	    nfds = linux_select(FD_SETSIZE, &ibits, &obits, &ebits,
+ 		 (sigisempty(&ndet_sigs_received)) ? timer : &ndet_polling_tv);
+ #endif
  #else SVR4
  	    nfds = notify_select(FD_SETSIZE, &ibits, &obits, &ebits,
  		(sigisempty(&ndet_sigs_received)) ? timer : &ndet_polling_tv);
***************
*** 401,407 ****
--- 414,422 ----
      FD_ZERO(&ndet_obits);
      FD_ZERO(&ndet_ebits);
      sigdelset( &ndet_sigs_auto, SIGIO );
+ #ifndef __linux
      sigdelset( &ndet_sigs_auto, SIGURG );
+ #endif
      /* Recompute all bits */
      (void) ntfy_new_enum_conditions(ntfy_cndtbl[(int) NTFY_INPUT],
  				    ndet_fd_change, NTFY_ENUM_DATA_NULL);
***************
*** 411,417 ****
--- 426,434 ----
  				    ndet_fd_change, NTFY_ENUM_DATA_NULL);
      /* Toggle notifier auto signal catching if situation changed */
      ndet_toggle_auto(&sigs_tmp, SIGIO);
+ #ifndef __linux
      ndet_toggle_auto(&sigs_tmp, SIGURG);
+ #endif
  }
  
  /* ARGSUSED */
***************
*** 715,721 ****
  		/*
  		 * Don't catch this signal, currently we are
  		 */
! #ifndef SVR4
  		n = sigvec(sig, &ndet_prev_sigvec[sig],
  			   (struct sigvec *) 0);	/* SYSTEM CALL */
  		ntfy_assert(n == 0, 6 /* Unexpected error: sigvec */);
--- 732,738 ----
  		/*
  		 * Don't catch this signal, currently we are
  		 */
! #if !defined(SVR4) && !defined(__linux)
  		n = sigvec(sig, &ndet_prev_sigvec[sig],
  			   (struct sigvec *) 0);	/* SYSTEM CALL */
  		ntfy_assert(n == 0, 6 /* Unexpected error: sigvec */);
***************
*** 743,749 ****
  	int             n;
  
  	/* Arrange to catch this signal, currently we are not */
! #ifndef SVR4
  	n = sigvec(sig, &ndet_sigvec, &ndet_prev_sigvec[sig]);
  	/* SYSTEM CALL */
  	ntfy_assert(n == 0, 8 /* Unexpected error: sigvec */);
--- 760,766 ----
  	int             n;
  
  	/* Arrange to catch this signal, currently we are not */
! #if !defined(SVR4) && !defined(__linux)
  	n = sigvec(sig, &ndet_sigvec, &ndet_prev_sigvec[sig]);
  	/* SYSTEM CALL */
  	ntfy_assert(n == 0, 8 /* Unexpected error: sigvec */);
***************
*** 768,774 ****
  ndet_signal_catcher(sig, code, scp)
      int             sig;
      int             code;
! #ifndef SVR4
      struct sigcontext *scp;
  #else SVR4
      ucontext_t *scp;
--- 785,791 ----
  ndet_signal_catcher(sig, code, scp)
      int             sig;
      int             code;
! #if !defined(SVR4) && !defined(__linux)
      struct sigcontext *scp;
  #else SVR4
      ucontext_t *scp;
***************
*** 775,781 ****
  #endif SVR4
  {
  
! #ifdef SVR4
      void        (*old_handler) () = ndet_prev_sigvec[sig].sa_handler;
  #else
      void        (*old_handler) () = ndet_prev_sigvec[sig].sv_handler;
--- 792,798 ----
  #endif SVR4
  {
  
! #if defined(SVR4) || defined(__linux)
      void        (*old_handler) () = ndet_prev_sigvec[sig].sa_handler;
  #else
      void        (*old_handler) () = ndet_prev_sigvec[sig].sv_handler;
***************
*** 1031,1037 ****
      return (ndet_signal_code);
  }
  
! #ifndef SVR4
  extern struct sigcontext *
  #else SVR4
  extern ucontext_t *
--- 1048,1054 ----
      return (ndet_signal_code);
  }
  
! #if !defined(SVR4) && !defined(__linux)
  extern struct sigcontext *
  #else SVR4
  extern ucontext_t *
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ndetsitimr.c xview3.2/lib//libxview/notify/ndetsitimr.c
*** xvorig/xview3.2/lib//libxview/notify/ndetsitimr.c	Tue Jun 29 08:18:06 1993
--- xview3.2/lib//libxview/notify/ndetsitimr.c	Thu Sep 16 23:05:51 1993
***************
*** 82,88 ****
       * Have notifier check for itimer changes next time around loop. Will
       * notice and adjust signal handling at this time.
       */
!     ndet_flags |= (type == NTFY_REAL_ITIMER) ? NDET_REAL_CHANGE :
  	NDET_VIRTUAL_CHANGE;
  Done:
      NTFY_END_CRITICAL;
--- 82,88 ----
       * Have notifier check for itimer changes next time around loop. Will
       * notice and adjust signal handling at this time.
       */
!     ndet_flags |= (type == NTFY_REAL_ITIMER) ?  NDET_REAL_CHANGE :
  	NDET_VIRTUAL_CHANGE;
  Done:
      NTFY_END_CRITICAL;
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ndis_d_pri.c xview3.2/lib//libxview/notify/ndis_d_pri.c
*** xvorig/xview3.2/lib//libxview/notify/ndis_d_pri.c	Tue Jun 29 08:18:01 1993
--- xview3.2/lib//libxview/notify/ndis_d_pri.c	Fri Aug 20 22:44:57 1993
***************
*** 16,21 ****
--- 16,24 ----
  #include <xview_private/ntfy.h>
  #include <xview_private/ndis.h>
  #include <signal.h>
+ #if defined(__linux) && !defined(NBBY)
+ #define NBBY 8
+ #endif
  
  typedef enum notify_error (*Notify_error_func) ();
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/notify.h xview3.2/lib//libxview/notify/notify.h
*** xvorig/xview3.2/lib//libxview/notify/notify.h	Tue Jun 29 08:18:15 1993
--- xview3.2/lib//libxview/notify/notify.h	Tue Oct  5 21:48:48 1993
***************
*** 22,27 ****
--- 22,28 ----
  #include <sys/time.h>
  #include <sys/resource.h>
  #include <xview/base.h>
+ #ifndef __linux
  #ifdef SYSV_WAIT
  #include <sys/rusage.h>
  #endif 
***************
*** 28,33 ****
--- 29,38 ----
  #ifdef SYSV_UCONTEXT
  #include <sys/ucontext.h>
  #endif 
+ #else /* __linux */
+ #include <sys/signal.h>  /* sigset_t */
+ typedef int ucontext_t;
+ #endif
  
  /*
   ***********************************************************************
***************
*** 277,283 ****
  
  #ifndef SYSV_UCONTEXT
  EXTERN_FUNCTION (struct sigcontext *notify_get_signal_context, (void));
! #else 
  EXTERN_FUNCTION (ucontext_t *notify_get_signal_context, (void));
  #endif
  
--- 282,288 ----
  
  #ifndef SYSV_UCONTEXT
  EXTERN_FUNCTION (struct sigcontext *notify_get_signal_context, (void));
! #else
  EXTERN_FUNCTION (ucontext_t *notify_get_signal_context, (void));
  #endif
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ntfy.h xview3.2/lib//libxview/notify/ntfy.h
*** xvorig/xview3.2/lib//libxview/notify/ntfy.h	Tue Jun 29 08:18:14 1993
--- xview3.2/lib//libxview/notify/ntfy.h	Sun Oct 10 12:37:13 1993
***************
*** 149,155 ****
   */
  typedef	struct ntfy_wait3_data {
  	int	pid;			/* Process waiting for */
! #ifndef SVR4
  	union	wait status;		/* Return value from wait3 */
  #else SVR4
  	int 	status;		/* Return value from wait3 */
--- 149,155 ----
   */
  typedef	struct ntfy_wait3_data {
  	int	pid;			/* Process waiting for */
! #if !defined(SVR4) && !defined(__linux)
  	union	wait status;		/* Return value from wait3 */
  #else SVR4
  	int 	status;		/* Return value from wait3 */
***************
*** 456,462 ****
  /*
   * Debugging aids.
   */
! #define	NTFY_DEBUG	1
  /*
   * Ntfy_set_errno is for setting notify_errno when there is really something
   * wrong.  An error message is displayed with notifier code has been compiled
--- 456,462 ----
  /*
   * Debugging aids.
   */
! #define	NTFY_DEBUG	0
  /*
   * Ntfy_set_errno is for setting notify_errno when there is really something
   * wrong.  An error message is displayed with notifier code has been compiled
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ntfy_fd_op.c xview3.2/lib//libxview/notify/ntfy_fd_op.c
*** xvorig/xview3.2/lib//libxview/notify/ntfy_fd_op.c	Tue Jun 29 08:18:03 1993
--- xview3.2/lib//libxview/notify/ntfy_fd_op.c	Thu Jul 15 22:33:06 1993
***************
*** 10,15 ****
--- 10,18 ----
   *	file for terms of the license.
   */
  
+ #ifdef __linux
+ #include <sys/time.h>
+ #endif
  #include <sys/types.h>
  #include <xview_private/ultrix_cpt.h>
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/ntfyclient.c xview3.2/lib//libxview/notify/ntfyclient.c
*** xvorig/xview3.2/lib//libxview/notify/ntfyclient.c	Tue Jun 29 08:18:11 1993
--- xview3.2/lib//libxview/notify/ntfyclient.c	Sat Jul 10 20:51:24 1993
***************
*** 18,24 ****
--- 18,26 ----
  #include <xview_private/ntfy.h>
  #include <xview_private/ndis.h>	/* For ndis_default_prioritizer */
  #include <xview_private/ndet.h>	
+ #ifndef __linux
  #include <search.h>
+ #endif
  #include <xview_private/portable.h>
  
  /* Variables used in paranoid enumerator (see ntfy_condition) */
***************
*** 29,34 ****
--- 31,42 ----
  static void *ndet_rootv;
  static void **ndet_root = &ndet_rootv;
  
+ /* This file seems to implement some kind of speed hack using b-trees
+  * to locate notifier clients. Linux does not have the search code in
+  * the libs, so I removed it in the hope that it keeps working, albeit
+  * slower. Of course it would be nice if someone added tsearch() & friends
+  * for linux (see <search.h> on Solaris). (lmfken) */
+ #ifndef __linux
  
  static int ndet_compar( key1, key2 )
     const void *key1, *key2;
***************
*** 45,50 ****
--- 53,59 ----
     key2a = tmp1 | ((key2a << 5) & 0x001f0000) | tmp;
     return key1a - key2a;
  }
+ #endif /* __linux */
  
  pkg_private NTFY_CLIENT *
  ntfy_find_nclient(client_list, nclient, client_latest)
***************
*** 60,65 ****
--- 69,75 ----
      if (*client_latest && (*client_latest)->nclient == nclient)
  	return (*client_latest);
  
+ #ifndef __linux
      if(( client_list == ndet_clients ) && ndet_clients ) {
         dummy_client.nclient = nclient;
         /* Find client */
***************
*** 75,80 ****
--- 85,91 ----
      }
  
      else 
+ #endif /* __linux */
          /* Search entire list */
          for (client = client_list; client; client = next) {
              next = client->next;
***************
*** 98,105 ****
      NTFY_CLIENT   **client_latest;
  {
      register NTFY_CLIENT *client;
!     static NTFY_CLIENT *new_client;
  
      if( client_list == &ndet_clients ) {
          if( new_client  == NTFY_CLIENT_NULL ) {
              if ((new_client = ntfy_alloc_client()) == NTFY_CLIENT_NULL)
--- 109,117 ----
      NTFY_CLIENT   **client_latest;
  {
      register NTFY_CLIENT *client;
!     static NTFY_CLIENT *new_client = NTFY_CLIENT_NULL;
  
+ #ifndef __linux
      if( client_list == &ndet_clients ) {
          if( new_client  == NTFY_CLIENT_NULL ) {
              if ((new_client = ntfy_alloc_client()) == NTFY_CLIENT_NULL)
***************
*** 120,126 ****
              return client;
      }
  
!     else if ((client = ntfy_find_nclient(*client_list, nclient,
  				    client_latest)) != NTFY_CLIENT_NULL)
          return client;
      /* Allocate client */
--- 132,140 ----
              return client;
      }
  
!     else
! #endif /* __linux */
!       if ((client = ntfy_find_nclient(*client_list, nclient,
  				    client_latest)) != NTFY_CLIENT_NULL)
          return client;
      /* Allocate client */
***************
*** 162,170 ****
--- 176,186 ----
  	next = condition->next;
  	ntfy_remove_condition(client, condition, who);
      }
+ #ifndef __linux
      /* Remove & free client from client_list */
      if( client_list == &ndet_clients )
          tdelete( client, ndet_root, ndet_compar );
+ #endif /* __linux */
      ntfy_remove_node((NTFY_NODE **) client_list, (NTFY_NODE *) client);
      /* Invalidate condition hint */
      *client_latest = NTFY_CLIENT_NULL;
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/sys_fcntl.c xview3.2/lib//libxview/notify/sys_fcntl.c
*** xvorig/xview3.2/lib//libxview/notify/sys_fcntl.c	Tue Jun 29 08:18:17 1993
--- xview3.2/lib//libxview/notify/sys_fcntl.c	Thu Jul 15 22:56:16 1993
***************
*** 15,21 ****
--- 15,25 ----
   */
  
  #ifndef SVR4
+ #ifndef __linux
  #include <syscall.h>
+ #else
+ #include "linux_select.h"
+ #endif
  #else SVR4
  #include <sys/syscall.h>
  #endif SVR4
***************
*** 25,29 ****
--- 29,37 ----
  notify_fcntl(fd, cmd, arg)
      int             fd, cmd, arg;
  {
+ #ifndef __linux
      return (syscall(SYS_fcntl, fd, cmd, arg));
+ #else
+     return (sys_fcntl(fd, cmd, arg));
+ #endif
  }
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/sys_read.c xview3.2/lib//libxview/notify/sys_read.c
*** xvorig/xview3.2/lib//libxview/notify/sys_read.c	Tue Jun 29 08:18:01 1993
--- xview3.2/lib//libxview/notify/sys_read.c	Thu Jul 15 23:01:46 1993
***************
*** 15,21 ****
--- 15,25 ----
   */
  
  #ifndef SVR4
+ #ifndef __linux
  #include <syscall.h>
+ #else
+ #include "linux_select.h"
+ #endif
  #else SVR4
  #include <sys/syscall.h>
  #endif SVR4
***************
*** 27,31 ****
--- 31,39 ----
      char           *buf;
      int             nbytes;
  {
+ #ifndef __linux
      return (syscall(SYS_read, fd, buf, nbytes));
+ #else
+     return (sys_read(fd, buf, (off_t)nbytes));
+ #endif
  }
diff -c -r --new-file xvorig/xview3.2/lib//libxview/notify/sys_select.c xview3.2/lib//libxview/notify/sys_select.c
*** xvorig/xview3.2/lib//libxview/notify/sys_select.c	Tue Jun 29 08:18:11 1993
--- xview3.2/lib//libxview/notify/sys_select.c	Sun Jul 18 03:33:11 1993
***************
*** 15,21 ****
--- 15,25 ----
   */
  
  #ifndef SVR4
+ #ifndef __linux
  #include <syscall.h>
+ #else
+ #include "linux_select.h"
+ #endif
  #else SVR4
  #include <values.h>
  #include <sys/time.h>
***************
*** 41,47 ****
--- 45,56 ----
  notify_select(nfds, in0, out0, ex0, tv)
  #endif SVR4
  #ifndef SVR4
+ #ifndef __linux
      int             nfds, *readfds, *writefds, *exceptfds;
+ #else
+     int             nfds;
+     fd_set *readfds, *writefds, *exceptfds;
+ #endif /* __linux */
  #else SVR4
      int nfds;
      fd_set *in0, *out0, *ex0;
***************
*** 50,59 ****
--- 59,76 ----
  {
  
  #ifndef SVR4
+ #ifndef __linux
      nfds = syscall(SYS_select, nfds, readfds, writefds, exceptfds, tv);
      ntfy_assert(!(nfds == 0 && tv == (struct timeval *) 0 &&
  		  *readfds == 0 && *writefds == 0 && *exceptfds == 0), 39
  		/* SYS_select returned when no stimuli */);
+ #else
+     nfds = linux_select(nfds, readfds, writefds, exceptfds, tv);
+     ntfy_assert(!(nfds == 0 && tv == (struct timeval *) 0 &&
+ 		readfds->fds_bits[0] == 0 && writefds->fds_bits[0] == 0 &&
+ 		exceptfds->fds_bits[0] == 0), 39
+ 		/* SYS_select returned when no stimuli */);
+ #endif
      return (nfds);
  #else SVR4
      /* register declarations ordered by expected frequency of use */
diff -c -r --new-file xvorig/xview3.2/lib//libxview/openwin/ow_view.c xview3.2/lib//libxview/openwin/ow_view.c
*** xvorig/xview3.2/lib//libxview/openwin/ow_view.c	Tue Jun 29 08:16:43 1993
--- xview3.2/lib//libxview/openwin/ow_view.c	Wed Aug  4 05:55:42 1993
***************
*** 101,107 ****
--- 101,115 ----
      }
  }
  
+ /* Linux: openwin_check_views() has turned static in 3.2, but is not
+  * called from anywhere. openwin_check_view() is made redundant at the same
+  * time. Is this a xview bug or can these routines be removed altogether??
+  * Change static to Pkg_private on linux, for now. */
+ #ifndef __linux
  static int
+ #else
+ Pkg_private int
+ #endif
  openwin_check_views(owin)
      Xv_openwin_info *owin;
  {
diff -c -r --new-file xvorig/xview3.2/lib//libxview/panel/p_list.c xview3.2/lib//libxview/panel/p_list.c
*** xvorig/xview3.2/lib//libxview/panel/p_list.c	Tue Jun 29 08:16:48 1993
--- xview3.2/lib//libxview/panel/p_list.c	Mon Aug 16 22:03:25 1993
***************
*** 3789,3795 ****
  
      /* weigh timeval's against multiclick-timeout resource */
      is_multiclick = panel_is_multiclick(ip->panel, 
! 					dp->last_click_row->click_time,
  					&event_time(event)
  					);
  
--- 3789,3795 ----
  
      /* weigh timeval's against multiclick-timeout resource */
      is_multiclick = panel_is_multiclick(ip->panel, 
! 					&dp->last_click_row->click_time,
  					&event_time(event)
  					);
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/panel/p_set.c xview3.2/lib//libxview/panel/p_set.c
*** xvorig/xview3.2/lib//libxview/panel/p_set.c	Tue Jun 29 08:16:59 1993
--- xview3.2/lib//libxview/panel/p_set.c	Sat Jul 10 20:53:32 1993
***************
*** 14,20 ****
  #include <xview/font.h>
  #include <xview/scrollbar.h>
  #include <xview/xv_xrect.h>
- #include <xview/font.h>
  #include <xview_private/draw_impl.h>
  
  Xv_private void	    win_set_no_focus();
--- 14,19 ----
diff -c -r --new-file xvorig/xview3.2/lib//libxview/panel/p_utl.c xview3.2/lib//libxview/panel/p_utl.c
*** xvorig/xview3.2/lib//libxview/panel/p_utl.c	Tue Jun 29 08:16:51 1993
--- xview3.2/lib//libxview/panel/p_utl.c	Sun Oct 10 21:53:36 1993
***************
*** 169,179 ****
--- 169,181 ----
              xv_free(image_string_wc(dest));
      }
  #else
+ #ifndef __linux
      {
          if (image_string(dest))
              xv_free(image_string(dest));
      }
  #endif
+ #endif
  
      size.x = size.y = 0;
      dest->im_type = type_code;
***************
*** 195,200 ****
--- 197,212 ----
  	    value_str = "";
  	if (!(str = (char *) panel_strsave((u_char *) value_str)))
  	    return (size);
+ #ifdef __linux
+ /* XView bug: This routine sometimes used a value that was already freed,
+  * leading to clobbered menu items. The problem is the 
+  * 'xv_free(image_string(dest))' above. In some cases the new 'value' 
+  * was the same memory area as 'image_string(dest)'. If free() modifies
+  * a freed buffer, the bug becomes visible. Kludgy fix for Linux. The
+  * real fix should probably be somewhere else. (lmfken Oct-93) */
+         if (image_string(dest))
+             xv_free(image_string(dest));
+ #endif
  	image_set_string(dest, str);
  #endif /* OW_I18N */
  	panel_image_set_font(dest, font);
diff -c -r --new-file xvorig/xview3.2/lib//libxview/panel/panel.c xview3.2/lib//libxview/panel/panel.c
*** xvorig/xview3.2/lib//libxview/panel/panel.c	Tue Jun 29 08:16:57 1993
--- xview3.2/lib//libxview/panel/panel.c	Tue Aug  3 06:01:36 1993
***************
*** 28,34 ****
--- 28,38 ----
  
  static int      panel_layout();
  
+ #ifndef __linux
  Xv_private_data Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #else
+ extern Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #endif
  
  /* default timer value */
  static struct itimerval PANEL_TIMER = {0, 500000, 0, 500000};
diff -c -r --new-file xvorig/xview3.2/lib//libxview/panel/panel_impl.h xview3.2/lib//libxview/panel/panel_impl.h
*** xvorig/xview3.2/lib//libxview/panel/panel_impl.h	Tue Jun 29 08:16:50 1993
--- xview3.2/lib//libxview/panel/panel_impl.h	Sun Jul 11 19:53:07 1993
***************
*** 10,16 ****
  #define panel_impl_defined
  
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 10,16 ----
  #define panel_impl_defined
  
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/scrollbar/sb.c xview3.2/lib//libxview/scrollbar/sb.c
*** xvorig/xview3.2/lib//libxview/scrollbar/sb.c	Tue Jun 29 08:17:06 1993
--- xview3.2/lib//libxview/scrollbar/sb.c	Tue Aug  3 06:02:27 1993
***************
*** 43,49 ****
--- 43,53 ----
  
  Xv_private void	win_set_no_focus();
  Xv_private Graphics_info	*xv_init_olgx();
+ #ifndef __linux
  Xv_private_data Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #else
+ extern Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #endif
  
  /******************************************************************/
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/sel/sel_agent.c xview3.2/lib//libxview/sel/sel_agent.c
*** xvorig/xview3.2/lib//libxview/sel/sel_agent.c	Mon Oct 11 01:12:48 1993
--- xview3.2/lib//libxview/sel/sel_agent.c	Sun Jul 11 17:42:51 1993
***************
*** 31,41 ****
   * Ultrix
   */
  #include <xview_private/ultrix_cpt.h>
! #ifdef SVR4
  #include <stdlib.h>
  #include <unistd.h>
  #endif SVR4
- 
  
  static void     selection_agent_process_functions();
  static Seln_result selection_agent_process_request();
--- 31,40 ----
   * Ultrix
   */
  #include <xview_private/ultrix_cpt.h>
! #if defined(SVR4) || defined(__linux)
  #include <stdlib.h>
  #include <unistd.h>
  #endif SVR4
  
  static void     selection_agent_process_functions();
  static Seln_result selection_agent_process_request();
diff -c -r --new-file xvorig/xview3.2/lib//libxview/sel/seln_impl.h xview3.2/lib//libxview/sel/seln_impl.h
*** xvorig/xview3.2/lib//libxview/sel/seln_impl.h	Tue Jun 29 08:15:25 1993
--- xview3.2/lib//libxview/sel/seln_impl.h	Sun Jul 11 19:49:13 1993
***************
*** 11,17 ****
  
  #include <errno.h>
  #ifndef FILE
! #ifndef SVR4
  #undef NULL
  #endif SVR4
  #include <stdio.h>
--- 11,17 ----
  
  #include <errno.h>
  #ifndef FILE
! #if !defined(SVR4) && !defined(__linux)
  #undef NULL
  #endif SVR4
  #include <stdio.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/selection/sel_impl.h xview3.2/lib//libxview/selection/sel_impl.h
*** xvorig/xview3.2/lib//libxview/selection/sel_impl.h	Tue Jun 29 08:16:02 1993
--- xview3.2/lib//libxview/selection/sel_impl.h	Tue Aug  3 05:33:39 1993
***************
*** 218,223 ****
--- 218,224 ----
  Xv_private void xv_sel_send_old_owner_sel_clear();
  Xv_private void xv_sel_set_compat_data();
  
+ #if !defined(__linux) || defined(__DEFINE_SEL_IMPL_VARS)
  XContext  selCtx;
  XContext  reqCtx;
  XContext  targetCtx;
***************
*** 224,229 ****
--- 225,238 ----
  XContext  propCtx;
  XContext  replyCtx;
  XContext  cmpatCtx;
+ #else
+ extern XContext  selCtx;
+ extern XContext  reqCtx;
+ extern XContext  targetCtx;
+ extern XContext  propCtx;
+ extern XContext  replyCtx;
+ extern XContext  cmpatCtx;
+ #endif
  
  #endif /* sel_impl_defined */
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/selection/sel_req.c xview3.2/lib//libxview/selection/sel_req.c
*** xvorig/xview3.2/lib//libxview/selection/sel_req.c	Mon Oct 11 01:12:48 1993
--- xview3.2/lib//libxview/selection/sel_req.c	Sun Jul 11 02:51:55 1993
***************
*** 17,23 ****
  #include <xview/notify.h>
  #include <sys/time.h>
  
! #ifdef SVR4 
  #include <stdlib.h> 
  #endif SVR4
  
--- 17,23 ----
  #include <xview/notify.h>
  #include <sys/time.h>
  
! #if defined(SVR4) || defined(__linux)
  #include <stdlib.h> 
  #endif SVR4
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/selection/selection.c xview3.2/lib//libxview/selection/selection.c
*** xvorig/xview3.2/lib//libxview/selection/selection.c	Tue Jun 29 08:16:03 1993
--- xview3.2/lib//libxview/selection/selection.c	Tue Aug  3 05:32:16 1993
***************
*** 10,16 ****
--- 10,23 ----
   *	file for terms of the license.
   */
  
+ #ifdef __linux
+ /* Kludge to prevent multiple variables with same name */
+ #define __DEFINE_SEL_IMPL_VARS
+ #endif
  #include <xview_private/sel_impl.h>
+ #ifdef __linux
+ #undef __DEFINE_SEL_IMPL_VARS
+ #endif
  #include <xview/window.h>
  
  Pkg_private char *xv_sel_atom_to_str(/* display, atom */);
diff -c -r --new-file xvorig/xview3.2/lib//libxview/server/server.c xview3.2/lib//libxview/server/server.c
*** xvorig/xview3.2/lib//libxview/server/server.c	Tue Jun 29 08:15:29 1993
--- xview3.2/lib//libxview/server/server.c	Tue Aug  3 05:55:16 1993
***************
*** 100,108 ****
--- 100,115 ----
  Xv_private_data char 	*xv_shell_prompt;
  
  /* global default server parameters */
+ #ifndef __linux
  Xv_private_data Xv_Server xv_default_server;
  Xv_private_data Xv_Screen xv_default_screen;
  Xv_private_data Display *xv_default_display;
+ #else
+ /* Globals already defined in base/xv_init.c */
+ extern Xv_Server xv_default_server;
+ extern Xv_Screen xv_default_screen;
+ extern Display *xv_default_display;
+ #endif
  
  /* Global data */
  Xv_private_data Defaults_pairs xv_kbd_cmds_value_pairs[4] = {
diff -c -r --new-file xvorig/xview3.2/lib//libxview/server/svr_kmdata.h xview3.2/lib//libxview/server/svr_kmdata.h
*** xvorig/xview3.2/lib//libxview/server/svr_kmdata.h	Tue Jun 29 08:15:30 1993
--- xview3.2/lib//libxview/server/svr_kmdata.h	Sun Jul 11 02:56:50 1993
***************
*** 14,20 ****
  
  #define MAX_NBR_MAPPINGS 6
  
! 
  
  /***************************************************************************
   *
--- 14,22 ----
  
  #define MAX_NBR_MAPPINGS 6
  
! #if defined(__linux) && !defined(NULL)
! #define NULL 0
! #endif
  
  /***************************************************************************
   *
diff -c -r --new-file xvorig/xview3.2/lib//libxview/server/svr_x.c xview3.2/lib//libxview/server/svr_x.c
*** xvorig/xview3.2/lib//libxview/server/svr_x.c	Tue Jun 29 08:15:27 1993
--- xview3.2/lib//libxview/server/svr_x.c	Tue Aug  3 05:59:35 1993
***************
*** 28,34 ****
--- 28,39 ----
  #include <X11/keysym.h>
  
  extern Display *XOpenDisplay();
+ #ifndef __linux
  Xv_private_data Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #else
+ /* Global already defined and initialized in in server/server.c */
+ extern Defaults_pairs xv_kbd_cmds_value_pairs[4];
+ #endif
  
  /*
   * The following table describes the default key mappings for special 
diff -c -r --new-file xvorig/xview3.2/lib//libxview/textsw/ei_text.c xview3.2/lib//libxview/textsw/ei_text.c
*** xvorig/xview3.2/lib//libxview/textsw/ei_text.c	Tue Jun 29 08:17:37 1993
--- xview3.2/lib//libxview/textsw/ei_text.c	Sat Aug 14 01:04:04 1993
***************
*** 105,111 ****
--- 105,116 ----
  #ifdef OW_I18N
  #define ISCNTRL(c)		((c >= 0) && iswcntrl(c))
  #else
+ #ifndef __linux
  #define ISCNTRL(c)		((128 <= c && c <= 159) || (iscntrl(c)))
+ #else
+ /* Linux: Allow 8-bit chars (do not treat them as control chars) */
+ #define ISCNTRL(c)		(c < 32)
+ #endif /* __linux */
  #endif
  
  Pkg_private Ei_handle ei_plain_text_create();
***************
*** 729,734 ****
--- 734,747 ----
  	    }
  	    batch++;
  	} else {
+ #ifdef __linux
+ /* Problems with 8-bit characters: 'aring' (0xe5) causes coredump in textedit
+  * because it's regarded as ISCNTRL(), leading to 'special_char 'getting a
+  * value > 255, leading to out-of-bound access of tempPf->pf_char[c]. Fix
+  * here by preventing too large special_char values.  */
+             if (c > 127)
+                 c -= 128;
+ #endif
  	    if (c == 127)
  		special_char = '?';
  	    else
diff -c -r --new-file xvorig/xview3.2/lib//libxview/textsw/es_file.c xview3.2/lib//libxview/textsw/es_file.c
*** xvorig/xview3.2/lib//libxview/textsw/es_file.c	Tue Jun 29 08:17:34 1993
--- xview3.2/lib//libxview/textsw/es_file.c	Sun Jul 11 18:10:45 1993
***************
*** 97,102 ****
--- 97,105 ----
  #include <sys/stat.h>
  #include <sys/file.h>
  #include <stdio.h>
+ #ifdef __linux
+ #include <unistd.h>
+ #endif
  #include <xview/pkg.h>
  #include <xview/attrol.h>
  #include <xview_private/primal.h>
***************
*** 338,343 ****
--- 341,349 ----
      char            name[MAXNAMLEN];      
      (void) wcstombs(name, name_wc, MAXNAMLEN);
  #endif /* OW_I18N */
+ #ifdef __linux
+     long int maxlinks;
+ #endif
  
      if (status == 0)
  	status = &dummy_status;
***************
*** 373,386 ****
--- 379,403 ----
  
  #ifndef BACKUP_AT_HEAD_OF_LINK
      /* (2) Chase the symbolic link if 'name' is one. */
+ #ifdef __linux
+     maxlinks = pathconf(name, _PC_LINK_MAX);
+ #endif
      for (temp_name = name, link_count = 0;
+ #ifndef __linux
  	 (link_count < MAXSYMLINKS) &&
+ #else
+ 	 (link_count < maxlinks) &&
+ #endif
  	 (-1 != (true_name_len =
  		 readlink(temp_name, true_name, sizeof(true_name))));
  	 temp_name = true_name, link_count++) {
  	true_name[true_name_len] = '\0';
      }
+ #ifndef __linux
      if (link_count == MAXSYMLINKS) {
+ #else
+     if (link_count == maxlinks) {
+ #endif
  	errno = ELOOP;
  	goto Error_Return;
      }
diff -c -r --new-file xvorig/xview3.2/lib//libxview/textsw/ev_edit.c xview3.2/lib//libxview/textsw/ev_edit.c
*** xvorig/xview3.2/lib//libxview/textsw/ev_edit.c	Tue Jun 29 08:17:35 1993
--- xview3.2/lib//libxview/textsw/ev_edit.c	Thu Aug 19 00:38:04 1993
***************
*** 497,508 ****
      switch (ev_xy_in_view(view, position, &lt_index, &rect)) {
        case EV_XY_BELOW:
  	/* BUG ALERT: The following heuristic must be replaced! */
  	delta = MIN(delta, position - line_seq[top_of_lc].pos);
  	if (delta < 50 * view->line_table.last_plus_one
  	    	&& lower_context < view->line_table.last_plus_one - 1
  	    	&& auto_scroll_by < view->line_table.last_plus_one - 1) {
- 
  	    old_top = line_seq[0].pos;
  	    new_top = ev_scroll_lines(view,
  				      MIN(view->line_table.last_plus_one - 1,
  					  MAX(1,
--- 497,520 ----
      switch (ev_xy_in_view(view, position, &lt_index, &rect)) {
        case EV_XY_BELOW:
  	/* BUG ALERT: The following heuristic must be replaced! */
+ #ifdef SCROLL_DEBUG
+         printf("delta=%d,pos=%d,lineseqpos=%d,last+one=%d,lcontext=%d,autoscrby=%d\n",
+                delta, position, line_seq[top_of_lc].pos,
+                view->line_table.last_plus_one, lower_context, auto_scroll_by);
+ #endif
+ #ifndef __linux
  	delta = MIN(delta, position - line_seq[top_of_lc].pos);
+ #endif
  	if (delta < 50 * view->line_table.last_plus_one
  	    	&& lower_context < view->line_table.last_plus_one - 1
  	    	&& auto_scroll_by < view->line_table.last_plus_one - 1) {
  	    old_top = line_seq[0].pos;
+ #ifdef SCROLL_DEBUG
+ 	printf("ev_scroll_lines(), MAX()=%d, total=%d\n", 
+ 	      MAX(1, MAX(lower_context, auto_scroll_by) + (delta / 50)),
+ 	      MIN(view->line_table.last_plus_one-1,MAX(1, MAX(lower_context, auto_scroll_by) + (delta / 50)))
+ 	);
+ #endif
  	    new_top = ev_scroll_lines(view,
  				      MIN(view->line_table.last_plus_one - 1,
  					  MAX(1,
***************
*** 524,529 ****
--- 536,545 ----
                      line_start = 0;
              }
  	    ev_set_start(view, line_start);
+ #ifdef SCROLL_DEBUG
+ 	printf("line_start=%d, lines_below=%d, lines_above=%d, scroll(2)\n",
+ 		line_start, lines_below, lines_above);
+ #endif
  	    new_top = ev_scroll_lines(view, 2, FALSE);
  	}
  	/*
***************
*** 532,537 ****
--- 548,557 ----
  	 */
          while ((old_top != new_top) &&
                 (position >= ev_index_for_line(view, top_of_lc))) {
+ #ifdef SCROLL_DEBUG
+ 	printf("old_top=%d, new_top=%d, position=%d ev_scroll_lines(2)\n",
+ 		old_top, new_top, position);
+ #endif
              old_top = new_top;
              new_top = ev_scroll_lines(view,
                          ((position - ev_index_for_line(view, top_of_lc))
diff -c -r --new-file xvorig/xview3.2/lib//libxview/textsw/txt_again.c xview3.2/lib//libxview/textsw/txt_again.c
*** xvorig/xview3.2/lib//libxview/textsw/txt_again.c	Tue Jun 29 08:17:32 1993
--- xview3.2/lib//libxview/textsw/txt_again.c	Sat Jul 24 20:57:53 1993
***************
*** 178,183 ****
--- 178,213 ----
   * Recording routines
   */
  
+ #ifdef __linux
+ /* The textsw_printf as written created a FILE struct, and called
+  * _doprnt() to format the text to an XView internal buffer. If I
+  * understand this correctly, it's just a replacement for vsprintf.
+  * Anyway, linux FILE isn't similar to the one assumed by the
+  * original code, so we'll have to do something.
+  * WARNING! This is untested!! */
+ 
+ static int
+ #ifdef ANSI_FUNC_PROTO
+ textsw_printf(register string_t *ptr_to_string, register char  *fmt, ...)
+ #else
+ textsw_printf(ptr_to_string, fmt, va_alist)
+     register string_t *ptr_to_string;
+     register char  *fmt;
+ va_dcl
+ #endif
+ {
+     va_list         args;
+     char *          sfree;
+     int             result;
+ 
+     VA_START(args, fmt);
+     sfree = (char *)TXTSW_STRING_FREE(ptr_to_string);
+     result = vsprintf(sfree, fmt, args);
+     va_end(args);
+     TXTSW_STRING_FREE(ptr_to_string) += strlen(sfree);
+     return (result);
+ }
+ #else /* __linux */
  /*
   * Following is stolen from 3.2ALPHA sprintf(str, fmt, va_alist) SIDE_EFFECT:
   * TXTSW_STRING_FREE(ptr_to_string) is modified by this routine.
***************
*** 324,329 ****
--- 354,360 ----
  }     
  
  #endif /* SVR4 */
+ #endif /* __linux */
  
  static	void
  textsw_record_buf(again, buffer, buffer_length)
***************
*** 613,618 ****
--- 644,728 ----
   * Replaying routines
   */
  
+ #ifdef __linux
+ /* The textsw_scanf as written created a FILE struct, and called
+  * _doscan() to do the scanning on the instring.
+  * This depends heavily on the internal structure of FILE, and is not
+  * applicable to linux. Instead I use a quickly hacked together
+  * scanf. The normal scanf isn't used because we need to get a
+  * pointer to where scanning of the instring stopped. I know no way
+  * to get that from sscanf(). */
+ 
+ static int
+ #ifdef ANSI_FUNC_PROTO
+ textsw_scanf(register string_t *ptr_to_string, register char  *fmt, ...)
+ #else
+ textsw_scanf(ptr_to_string, fmt, va_alist)
+     register string_t *ptr_to_string;
+     register char  *fmt;
+ va_dcl
+ #endif
+ {
+     va_list         args;
+     char *          sbase;
+     int             result;
+     char *f, *sp;
+     int c, *ip;
+ 
+     VA_START(args, fmt);
+     sbase = (char *)TXTSW_STRING_BASE(ptr_to_string);
+ 
+ /* scanf loop, supports %s, %15s, %d, %x because those are the types used
+  * by routines calling this function (textsw_scanf is static). */ 
+     result = 0;
+     for (f = fmt; *f; ++f) {
+       switch (*f) {
+       case '%':
+         ++f;
+         c = 999;              
+         if (isdigit(*f)) {
+           c = atoi(f);
+           while (isdigit(*f))
+             ++f;
+         }
+         for (; isspace(*sbase); ++sbase)
+           ;
+         switch (*f) {
+         case 's':
+           sp = va_arg(args, char *);
+           for (; *sbase && !isspace(*sbase) && c; ++sbase, --c) 
+             *sp++ = *sbase;
+           *sp = '\0';
+           ++result;
+           break;
+         case 'd':
+           ip = va_arg(args, int *);
+           if (isdigit(*sbase)) {
+             ++result;
+             *ip = atoi(sbase);
+             while (isdigit(*sbase))
+               ++sbase;
+           }
+           break;
+         case 'x':
+           ip = va_arg(args, int *);
+           result += sscanf(sbase, "%x", ip);
+           while (isxdigit(*sbase))
+             ++sbase;
+           break;
+         default:
+           break;
+         }
+         break;
+       default:
+         break;
+       }
+     }
+     va_end(args);
+     TXTSW_STRING_BASE(ptr_to_string) = sbase;
+     return (result);
+ }
+ #else /* __linux */
  /*
   * Following is stolen from sscanf(str, fmt, args) SIDE_EFFECT:
   * TXTSW_STRING_BASE(ptr_to_string) is modified by this routine.
***************
*** 705,710 ****
--- 815,821 ----
  
  #endif /* OW_I18N */
  }
+ #endif /* __linux */
  
  static int
  textsw_next_is_delimiter(again)
diff -c -r --new-file xvorig/xview3.2/lib//libxview/textsw/txt_edit.c xview3.2/lib//libxview/textsw/txt_edit.c
*** xvorig/xview3.2/lib//libxview/textsw/txt_edit.c	Tue Jun 29 08:17:36 1993
--- xview3.2/lib//libxview/textsw/txt_edit.c	Tue Aug  3 05:46:47 1993
***************
*** 28,34 ****
--- 28,39 ----
  #define UPDATE_SCROLLBAR(_delta, _old_length)\
  	((THRESHOLD * _delta) >= _old_length)
  
+ #ifndef __linux
  Xv_private_data char *xv_shell_prompt;
+ #else
+ /* Global already defined in server/server.c */
+ extern char *xv_shell_prompt;
+ #endif
  
  Pkg_private void     textsw_notify_replaced();
  Pkg_private Seln_rank textsw_acquire_seln();
diff -c -r --new-file xvorig/xview3.2/lib//libxview/textsw/txt_filter.c xview3.2/lib//libxview/textsw/txt_filter.c
*** xvorig/xview3.2/lib//libxview/textsw/txt_filter.c	Tue Jun 29 08:17:39 1993
--- xview3.2/lib//libxview/textsw/txt_filter.c	Wed Jul 14 01:11:33 1993
***************
*** 903,909 ****
--- 903,913 ----
  #ifdef SVR4
      if (xv_fcntl(to_filter[OUTPUT], F_SETFL, FNDELAY) == -1)
  #else
+ #if !defined(__linux) || defined(FNDELAY)
      if (fcntl(to_filter[OUTPUT], F_SETFL, FNDELAY) == -1)
+ #else
+     if (fcntl(to_filter[OUTPUT], F_SETFL, O_NONBLOCK) == -1)
+ #endif
  #endif
  	return (FAIL);
  
***************
*** 910,916 ****
--- 914,924 ----
  #ifdef SVR4
      if (xv_fcntl(from_filter[INPUT], F_SETFL, FNDELAY) == -1)
  #else
+ #if !defined(__linux) || defined(FNDELAY)
      if (fcntl(from_filter[INPUT], F_SETFL, FNDELAY) == -1)
+ #else
+     if (fcntl(from_filter[INPUT], F_SETFL, O_NONBLOCK) == -1)
+ #endif
  #endif
  	return (FAIL);
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/charscreen.h xview3.2/lib//libxview/ttysw/charscreen.h
*** xvorig/xview3.2/lib//libxview/ttysw/charscreen.h	Tue Jun 29 08:17:13 1993
--- xview3.2/lib//libxview/ttysw/charscreen.h	Tue Aug  3 05:25:21 1993
***************
*** 23,28 ****
--- 23,29 ----
   * Character dimensions (fixed width fonts only!)
   * and of screen in pixels.
   */
+ #if !defined(__linux) || defined(__DEFINE_CHARSCREEN_VARS)
  int	chrheight, chrwidth, chrbase;
  int	winheightp, winwidthp;
  int	chrleftmargin;
***************
*** 35,40 ****
--- 36,48 ----
   * and turn delaypainting off.
   */
  int	delaypainting;
+ #else /* __linux && !__DEFINE_CHARSCREEN_VARS */
+ extern int	chrheight, chrwidth, chrbase;
+ extern int	winheightp, winwidthp;
+ extern int	chrleftmargin;
+ extern struct	pixfont *pixfont;
+ extern int	delaypainting;
+ #endif
  
  #ifdef cplus
  void	ttysw_pstring(char *s, int col, int row);
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/csr_change.c xview3.2/lib//libxview/ttysw/csr_change.c
*** xvorig/xview3.2/lib//libxview/ttysw/csr_change.c	Tue Jun 29 08:17:16 1993
--- xview3.2/lib//libxview/ttysw/csr_change.c	Tue Aug  3 05:45:56 1993
***************
*** 44,50 ****
--- 44,55 ----
  #include <xview/attrol.h>
  #include <xview/server.h>
  #include <xview/font.h>
+ #ifndef __linux
  Xv_private_data char *xv_shell_prompt;
+ #else
+ /* Global already defined in server/server.c */
+ extern char *xv_shell_prompt;
+ #endif
  
  #ifdef OW_I18N
  #define NULL_CHARP      (CHAR *) 0
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/csr_init.c xview3.2/lib//libxview/ttysw/csr_init.c
*** xvorig/xview3.2/lib//libxview/ttysw/csr_init.c	Tue Jun 29 08:17:13 1993
--- xview3.2/lib//libxview/ttysw/csr_init.c	Tue Aug  3 05:23:14 1993
***************
*** 31,37 ****
--- 31,47 ----
  #include <xview/window.h>
  #include <xview/font.h>
  #include <xview_private/charimage.h>
+ #ifdef __linux
+ /* charscreen.h defines some variables, and then the header file is included
+  * by more than one source file. The linux shlib-tools don't like it.
+  * Kludge around it by setting a define here which makes the variable
+  * definitions visible, but do NOT set it in the other source files. */  
+ #define __DEFINE_CHARSCREEN_VARS
+ #endif
  #include <xview_private/charscreen.h>
+ #ifdef __linux
+ #undef __DEFINE_CHARSCREEN_VARS
+ #endif
  #ifdef OW_I18N
  #ifdef FULL_R5
  #include <xview_private/term_impl.h>
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/term_ntfy.c xview3.2/lib//libxview/ttysw/term_ntfy.c
*** xvorig/xview3.2/lib//libxview/ttysw/term_ntfy.c	Tue Jun 29 08:17:22 1993
--- xview3.2/lib//libxview/ttysw/term_ntfy.c	Mon Aug 30 19:21:31 1993
***************
*** 63,69 ****
  /* performance: global cache of getdtablesize() */
  extern int      dtablesize_cache;
  
! #ifdef SVR4
  #define GETDTABLESIZE() \
  (dtablesize_cache?dtablesize_cache:(dtablesize_cache=(int)sysconf(_SC_OPEN_MAX)))
  #else
--- 63,69 ----
  /* performance: global cache of getdtablesize() */
  extern int      dtablesize_cache;
  
! #if defined(SVR4) || defined(__linux)
  #define GETDTABLESIZE() \
  (dtablesize_cache?dtablesize_cache:(dtablesize_cache=(int)sysconf(_SC_OPEN_MAX)))
  #else
***************
*** 148,155 ****
          action |= 0x80;
      }
  
- 
- 
      if ((action == ACTION_MENU) && down_event) {
  	/*
  	 * The view that should be affected by the menu is the one that the
--- 148,153 ----
***************
*** 173,181 ****
--- 171,181 ----
       * again, this is the place to start looking.
       */
      if (ttysw->pending_remote != ttysw->remote) {
+ #if !defined(__linux) || defined(TIOCREMOTE)
  	if (ioctl(ttysw->ttysw_pty, TIOCREMOTE, &ttysw->pending_remote) < 0)
  	    perror("ioctl: TIOCREMOTE");
  	else
+ #endif
  	    ttysw->remote = ttysw->pending_remote;
      }
  
***************
*** 193,198 ****
--- 193,199 ----
  	}
  	return (nv);
      }
+ 
      if (termsw->cooked_echo && down_event) {
  	insert = (int) xv_get(textsw, TEXTSW_INSERTION_POINT_I18N);
  	length = (int) xv_get(textsw, TEXTSW_LENGTH_I18N);
***************
*** 200,206 ****
--- 201,211 ----
  	    /*
  	     * Process pending literal next insertion at end of buffer.
  	     */
+ #ifndef __linux
  	    if (termsw->literal_next && action <= ASCII_LAST &&
+ #else
+ 	    if (termsw->literal_next && action <= ISO_LAST &&
+ #endif
  		    insert == length) {
  		CHAR	input_char = (CHAR) action;
  
***************
*** 236,242 ****
      if (!termsw->cooked_echo
  	&& (action == '\r' || action == '\n')) {
  
- 
          if (!tty_iscanon(ttysw)) /*1030878*/
              xv_set( textsw, TEXTSW_DIFFERENTIATE_CR_LF, 1, 0 );
  
--- 241,246 ----
***************
*** 274,283 ****
--- 278,289 ----
  	 *	should get the same treatment.  For a third, whatever tests we
  	 *	make should also apply to the cooked_echo case.
  	 */
+ #ifndef __linux
  	if (action == tty_getintrc(ttysw)) {
  	        (void) xv_set(textsw, TEXTSW_INSERTION_POINT_I18N,
  			  TEXTSW_INFINITY, 0);
  	}
+ #endif
  	(void) textsw_insert(textsw, &input_char, (long) 1);
  	nv = NOTIFY_DONE;
      } else if (termsw->cooked_echo && down_event) {
***************
*** 293,299 ****
--- 299,309 ----
  					(Notify_event) (event), arg, type);
  	}
      } else if (!termsw->cooked_echo && down_event &&
+ #ifndef __linux
  		action >= ASCII_FIRST && action <= ASCII_LAST) {
+ #else
+ 		action >= ASCII_FIRST && action <= ISO_LAST) {
+ #endif
  	/*
  	 * Pass the event through to the pty, storing it in the ttysw input
  	 * buffer until the notifier triggers code to feed it into the pty
***************
*** 361,368 ****
--- 371,384 ----
  	* Even if cooked echo is off, we still have to handle
  	* keyboard signals.
  	*/
+ #ifndef __linux
  	if (!termsw->cooked_echo && !tty_issig(ttysw))
  	    break;
+ #else
+ /* Linux: respect noncanonical mode, do not send interrupts */
+         if (!tty_issig(ttysw) || !tty_iscanon(ttysw))
+ 	    break;
+ #endif
  
  	/* Only look at the down event for control key */
  	if (event_is_up(event))
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/termsw.c xview3.2/lib//libxview/ttysw/termsw.c
*** xvorig/xview3.2/lib//libxview/ttysw/termsw.c	Tue Jun 29 08:17:15 1993
--- xview3.2/lib//libxview/ttysw/termsw.c	Thu Sep 30 23:10:09 1993
***************
*** 413,419 ****
--- 413,421 ----
  
      /* Set the PTY to operate as a "remote terminal". */
      fd = (int) xv_get(termsw_public, TTY_PTY_FD);
+ #if !defined(__linux) || defined(TIOCREMOTE)
      (void) ioctl(fd, TIOCREMOTE, &on);
+ #endif
      ttysw_folio->remote = ttysw_folio->pending_remote = on;
  
      /*
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_gtty.c xview3.2/lib//libxview/ttysw/tty_gtty.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_gtty.c	Tue Jun 29 08:17:20 1993
--- xview3.2/lib//libxview/ttysw/tty_gtty.c	Sun Jul 11 18:12:06 1993
***************
*** 34,39 ****
--- 34,83 ----
   */
  #ifdef	XV_USE_TERMIOS
  
+ #if defined(__linux) && !defined(CINTR)
+ /* The following values have been obtained from /usr/include/linux/tty.h.
+  * They represent the default tty modes on linux.
+ 	intr=^C		quit=^|		erase=del	kill=^U
+ 	eof=^D		vtime=\0	vmin=\1		sxtc=\0
+ 	start=^Q	stop=^S		susp=^Z		eol=\0
+ 	reprint=^R	discard=^U	werase=^W	lnext=^V
+ 	eol2=\0
+ */
+ /* This belongs in sys/termios.h, not here. */
+ #define CINTR 003
+ #define CQUIT 034
+ #define CERASE 0177
+ #define CKILL 025
+ #define CEOF 004
+ #define CTIME 0
+ #define CMIN 1
+ #define CSWTC 0
+ #define CSTART 021
+ #define CSTOP 023
+ #define CSUSP 032
+ #define CEOL 0
+ #define CREPRINT 022
+ #define CDISCARD 017
+ #define CWERASE 027
+ #define CLNEXT 026
+ #define CEOL2 0
+ #endif /* __linux && !CINTR */
+ 
+ #ifdef __linux
+ static struct termios   default_modes = {
+ 	ICRNL|IXON,	  		  	/* input modes */
+ 	OPOST|ONLCR,			    	/* output modes */
+ 	B38400|CS8|CREAD,   		 	/* control modes */
+ 	ISIG|ICANON|ECHO|ECHOCTL|ECHOKE, 	/* local modes */
+ 
+ 	CINTR, CQUIT, CERASE, CKILL,
+ 	CEOF, CTIME, CMIN, CSWTC,
+ 	CSTART, CSTOP, CSUSP, CEOL,
+ 	CREPRINT, CDISCARD, CWERASE, CLNEXT,
+ 	CEOL2
+ };
+ 	
+ #else /* __linux */
  static struct termios	default_modes = {
  	BRKINT|ICRNL|IXON|IGNPAR|IMAXBEL,	    	/* input modes */
  	OPOST|ONLCR,				    	/* output modes */
***************
*** 57,62 ****
--- 101,107 ----
  	CWERASE,	/* VWERASE */
  	CLNEXT,		/* VLNEXT */
  };
+ #endif /* __linux */
  
  #else	/* XV_USE_TERMIOS */
  
***************
*** 222,228 ****
--- 267,277 ----
      struct termios	*tp;
  {
      char	str[WE_TTYPARMSLEN];
+ #ifndef __linux
      short	temps[16];
+ #else
+     short	temps[17];
+ #endif
  
      if (_we_setstrfromenvironment(WE_TTYPARMS, str))
  	return (-1);
***************
*** 229,234 ****
--- 278,284 ----
      else {
  	register int i;
  
+ #ifndef __linux
  	if (sscanf(str,
  		"%ld,%ld,%ld,%ld,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd",
  		&tp->c_iflag, &tp->c_oflag, &tp->c_cflag, &tp->c_lflag,
***************
*** 246,251 ****
--- 296,320 ----
  	(void) putenv(WE_TTYPARMS_E);
  	return (0);
      }
+ #else /* __linux */
+ 	if (sscanf(str,
+ 		"%ld,%ld,%ld,%ld,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd",
+ 		&tp->c_iflag, &tp->c_oflag, &tp->c_cflag, &tp->c_lflag,
+ 		&temps[0],  &temps[1],  &temps[2],  &temps[3],
+ 		&temps[4],  &temps[5],  &temps[6],  &temps[7],
+ 		&temps[8],  &temps[9],  &temps[10], &temps[11],
+ 		&temps[12], &temps[13], &temps[14], &temps[15],
+ 		&temps[16]) != 21)
+ 	    return (-1);
+ 	for (i = 0; i <= VEOL2; i++)
+ 	    tp->c_cc[i] = temps[i];
+ 	/*
+ 	 * Always clear
+ 	 */
+ 	(void) putenv(WE_TTYPARMS_E);
+ 	return (0);
+     }
+ #endif /* __linux */
  }
  
  #else	/* XV_USE_TERMIOS */
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_impl.h xview3.2/lib//libxview/ttysw/tty_impl.h
*** xvorig/xview3.2/lib//libxview/ttysw/tty_impl.h	Tue Jun 29 08:17:18 1993
--- xview3.2/lib//libxview/ttysw/tty_impl.h	Mon Aug 30 19:57:19 1993
***************
*** 211,217 ****
--- 211,221 ----
   */
  #ifdef	XV_USE_TERMIOS
  #define	tty_gettabs(t)		((t)->termios.c_oflag & XTABS)
+ #if !defined(__linux) || defined(VDSUSP)
  #define	tty_getdsuspc(t)	((int) ((t)->termios.c_cc[VDSUSP]))
+ #else
+ #define	tty_getdsuspc(t)	((int) -1)
+ #endif
  #define	tty_geteofc(t)		((int) ((t)->termios.c_cc[VEOF]))
  #define	tty_geteolc(t)		((int) ((t)->termios.c_cc[VEOL]))
  #define	tty_geteol2c(t)		((int) ((t)->termios.c_cc[VEOL2]))
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_init.c xview3.2/lib//libxview/ttysw/tty_init.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_init.c	Tue Jun 29 08:17:19 1993
--- xview3.2/lib//libxview/ttysw/tty_init.c	Tue Jul 13 23:43:46 1993
***************
*** 408,414 ****
--- 408,418 ----
      if ((fdflags = fcntl(fd, F_GETFL, 0)) == -1)
  #endif
  	return (-1);
+ #if !defined(__linux) || defined(FNDELAY)
      fdflags |= FNDELAY;
+ #else
+     fdflags |= O_NONBLOCK;
+ #endif
  #ifdef SVR4
      if (xv_fcntl(fd, F_SETFL, fdflags) == -1)
  #else
***************
*** 432,442 ****
      char	    appname[20];
      char	    *p;
      unsigned        ttysw_error_sleep = 1;
! #ifndef SVR4
      struct sigvec   vec, ovec;
  #else
      struct sigaction	vec, ovec;
      extern char *ptsname();
  
  #define BSD_TTY_COMPAT /* yank this if csh ever gets ported properly */
  
--- 436,448 ----
      char	    appname[20];
      char	    *p;
      unsigned        ttysw_error_sleep = 1;
! #if !defined(SVR4) && !defined(__linux)
      struct sigvec   vec, ovec;
  #else
      struct sigaction	vec, ovec;
+ #ifndef __linux
      extern char *ptsname();
+ #endif
  
  #define BSD_TTY_COMPAT /* yank this if csh ever gets ported properly */
  
***************
*** 461,467 ****
      }
  
      /* Set up the child characteristics */
! #ifndef SVR4  	/* SunOS4.x code */
      vec.sv_handler = SIG_DFL;
      vec.sv_mask = vec.sv_onstack = 0;
      sigvec(SIGWINCH, &vec, 0);
--- 467,473 ----
      }
  
      /* Set up the child characteristics */
! #if !defined(SVR4) && !defined(__linux)  	/* SunOS4.x code */
      vec.sv_handler = SIG_DFL;
      vec.sv_mask = vec.sv_onstack = 0;
      sigvec(SIGWINCH, &vec, 0);
***************
*** 497,506 ****
--- 503,518 ----
      vec.sa_flags = SA_RESTART;
      sigaction(SIGTTOU, &vec, &ovec);
  
+ #ifndef __linux
      if (unlockpt(ttysw->ttysw_pty) == -1)
          perror("unlockpt (2)");
      if ((ttysw->ttysw_tty = open(ptsname(ttysw->ttysw_pty),O_RDWR))<0)
          return -1;
+ #else
+     setpgrp();
+     if ((ttysw->ttysw_tty = open(ttysw->tty_name,O_RDWR))<0)
+         return -1;
+ #endif
  
      sigaction(SIGTTOU, &ovec, (struct sigaction *) 0);
  #endif /* SVR4 */
***************
*** 538,544 ****
  	    offset++;
      }
  
! #ifdef SVR4
  #ifdef BSD_TTY_COMPAT
  /*
   * ttcompat seems to leave things in a funny state and assumes
--- 550,556 ----
  	    offset++;
      }
  
! #if defined(SVR4) || defined(__linux)
  #ifdef BSD_TTY_COMPAT
  /*
   * ttcompat seems to leave things in a funny state and assumes
***************
*** 934,940 ****
   */
  
  #ifndef SVR4
! 
  /*
   * Make entry in /etc/utmp for ttyfd. Note: this is dubious usage of
   * /etc/utmp but many programs (e.g. sccs) look there when determining who is
--- 946,952 ----
   */
  
  #ifndef SVR4
! #ifndef __linux
  /*
   * Make entry in /etc/utmp for ttyfd. Note: this is dubious usage of
   * /etc/utmp but many programs (e.g. sccs) look there when determining who is
***************
*** 1018,1022 ****
      }
      return (ttyslotuse);
  }
! #endif
  
--- 1030,1097 ----
      }
      return (ttyslotuse);
  }
! #else /* __linux */
! /* Linux version of updateutmp uses the getutXX functions, we don't
!  * need no ttyslot() or direct writing to /etc/utmp */
! int
! updateutmp(username, ttyslotuse, ttyfd)
!     char           *username;
!     int             ttyslotuse, ttyfd;
! {
!     /*
!      * Update /etc/utmp
!      */
!     struct utmp     utmp;
!     struct passwd  *passwdent;
!     int             f;
!     char           *ttyn;
!     extern char    *ttyname();
! 
!     memset(&utmp, 0, sizeof(utmp));
!     utmp.ut_type = USER_PROCESS;
!     if (!username) {
! 	/*
! 	 * Get login name
! 	 */
! 	if ((passwdent = getpwuid(getuid())) == (struct passwd *) NULL) {
! 	    (void) fprintf(stderr, 
! 		XV_MSG("couldn't find user name\n"));
! 	    return (0);
! 	}
! 	username = passwdent->pw_name;
!         utmp.ut_pid = getpid();
!     }
!     else
!       if (!(*username))
!         utmp.ut_type = DEAD_PROCESS; /* "" as username, logging out */
! 
!     utmp.ut_name[0] = '\0';	/* Set incase *username is 0 */
!     (void) strncpy(utmp.ut_name, username, sizeof(utmp.ut_name));
!     /*
!      * Get line (tty) name
!      */
!     ttyn = ttyname(ttyfd);
!     if (ttyn == (char *) NULL)
! 	ttyn = "/dev/tty??";
!     (void) strncpy(utmp.ut_line, XV_RINDEX(ttyn, '/') + 1, sizeof(utmp.ut_line));
!     (void) strncpy(utmp.ut_id, ttyn + strlen(ttyn) - 2, 2);
!     /*
!      * Set host to be empty
!      */
!     (void) strncpy(utmp.ut_host, "", sizeof(utmp.ut_host));
!     /*
!      * Get start time
!      */
!     (void) time((time_t *) (&utmp.ut_time));
!     /*
!      * Put utmp in /etc/utmp
!      */
!     setutent();
!     (void)getutline(&utmp);
!     pututline(&utmp);
!     endutent();
!     return 1;        /* Return dummy value for ttyslot number */
! }
! #endif /* __linux */
! #endif SVR4
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_main.c xview3.2/lib//libxview/ttysw/tty_main.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_main.c	Tue Jun 29 08:17:22 1993
--- xview3.2/lib//libxview/ttysw/tty_main.c	Mon Aug 30 19:57:02 1993
***************
*** 184,191 ****
  	    if (termsw->pty_eot > -1)
  		return (1);
      }
- 
- 
      return (0);
  }
  
--- 184,189 ----
***************
*** 906,913 ****
--- 904,915 ----
  
  	    if (int_ucntl == (tiocsti & 0xff))
  		ttysw_process_STI(ttysw, owbp, cc - 1);
+ #ifndef XV_USE_TERMIOS
  	    (void) ioctl(ttysw->ttysw_tty, TIOCGETC, &ttysw->tchars);
  	    (void) ioctl(ttysw->ttysw_tty, TIOCGLTC, &ttysw->ltchars);
+ #else
+             (void)tcgetattr(ttysw->ttysw_tty, &ttysw->termios);
+ #endif
  	    ttysw_getp(TTY_VIEW_HANDLE_FROM_TTY_FOLIO(ttysw));	/* jcb for nng */
  	} else
  #ifdef OW_I18N
***************
*** 954,959 ****
--- 956,965 ----
  	    ttysw->implicit_commit = 0;		/* turn off the flag */
  #endif
  	}
+ #ifdef DEBUG
+ 	printf("ttysw_consume_output(), calling ttysw_output_it() %d <%s>\n",
+ 		owbp - orbp, orbp);
+ #endif
  	cc = ttysw_output_it(ttysw_view, orbp, owbp - orbp);
  #ifdef OW_I18N
          /*
***************
*** 1511,1517 ****
       * Otherwise, we use the winsize struct  and TIOCSWINSZ ioctl.
       */
      struct winsize  ws;
! #ifndef SVR4
      struct sigvec vec, ovec;
  
      vec.sv_handler = SIG_IGN;
--- 1517,1523 ----
       * Otherwise, we use the winsize struct  and TIOCSWINSZ ioctl.
       */
      struct winsize  ws;
! #if !defined(SVR4) && !defined(__linux)
      struct sigvec vec, ovec;
  
      vec.sv_handler = SIG_IGN;
***************
*** 1524,1530 ****
      if ((ioctl(ttysw->ttysw_tty, TIOCSWINSZ, &ws)) == -1)
  	perror(XV_MSG("ttysw-TIOCSWINSZ"));
  
! #ifndef SVR4
      (void) sigvec(SIGTTOU, &ovec, 0);
  #endif
  #endif /* sun */
--- 1530,1536 ----
      if ((ioctl(ttysw->ttysw_tty, TIOCSWINSZ, &ws)) == -1)
  	perror(XV_MSG("ttysw-TIOCSWINSZ"));
  
! #if !defined(SVR4) && !defined(__linux)
      (void) sigvec(SIGTTOU, &ovec, 0);
  #endif
  #endif /* sun */
***************
*** 1624,1630 ****
  ttysw_flush_input(ttysw)
      Ttysw_folio     ttysw;
  {
! #ifndef SVR4
      struct sigvec   vec, ovec;	/* Sys V compatibility */
      int flushf = 0;
  
--- 1630,1636 ----
  ttysw_flush_input(ttysw)
      Ttysw_folio     ttysw;
  {
! #if !defined(SVR4) && !defined(__linux)
      struct sigvec   vec, ovec;	/* Sys V compatibility */
      int flushf = 0;
  
***************
*** 1655,1661 ****
  #   endif /* XV_USE_TERMIOS */
  	perror(XV_MSG("TIOCFLUSH"));
  
! #ifndef SVR4
      (void) sigvec(SIGTTOU, &ovec, (struct sigvec *) 0);
  #else
      sigaction(SIGTTOU, &ovec, (struct sigaction *) 0);
--- 1661,1667 ----
  #   endif /* XV_USE_TERMIOS */
  	perror(XV_MSG("TIOCFLUSH"));
  
! #if !defined(SVR4) && !defined(__linux)
      (void) sigvec(SIGTTOU, &ovec, (struct sigvec *) 0);
  #else
      sigaction(SIGTTOU, &ovec, (struct sigaction *) 0);
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_mapkey.c xview3.2/lib//libxview/ttysw/tty_mapkey.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_mapkey.c	Tue Jun 29 08:17:20 1993
--- xview3.2/lib//libxview/ttysw/tty_mapkey.c	Sat Jul 17 23:28:51 1993
***************
*** 500,506 ****
   * have more time.
   */
  
! #ifdef i386
  static void
  ttysw_arrow_keys_to_string(xv_id, str)
      unsigned        xv_id;
--- 500,506 ----
   * have more time.
   */
  
! #if defined(i386) && !defined(__linux)
  static void
  ttysw_arrow_keys_to_string(xv_id, str)
      unsigned        xv_id;
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_modes.c xview3.2/lib//libxview/ttysw/tty_modes.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_modes.c	Tue Jun 29 08:17:23 1993
--- xview3.2/lib//libxview/ttysw/tty_modes.c	Tue Aug 24 21:03:41 1993
***************
*** 125,133 ****
--- 125,135 ----
  	/*
  	 * Switch the pty out of remote mode.
  	 */
+ #if !defined(__linux) || defined(TIOCREMOTE)
  	if (ioctl(ttysw->ttysw_pty, TIOCREMOTE, &off) < 0)
  	    perror("ioctl: TIOCREMOTE");
  	else
+ #endif
  	    ttysw->remote = ttysw->pending_remote = off;
      }
  
***************
*** 276,284 ****
--- 278,288 ----
       */
      ttysw_getp((Ttysw_view_handle) ttysw_view);
      ttysw->pending_remote = termsw->cooked_echo;
+ #if !defined(__linux) || defined(TIOCREMOTE)
      if (ioctl(ttysw->ttysw_pty, TIOCREMOTE, &ttysw->pending_remote) < 0)
  	perror("ioctl: TIOCREMOTE");
      else
+ #endif
  	ttysw->remote = ttysw->pending_remote;
  
      if (!ttysw_waiting_for_pty_input) {
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_ntfy.c xview3.2/lib//libxview/ttysw/tty_ntfy.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_ntfy.c	Tue Jun 29 08:17:17 1993
--- xview3.2/lib//libxview/ttysw/tty_ntfy.c	Sat Aug 21 19:44:22 1993
***************
*** 179,185 ****
--- 179,189 ----
       * Only killpg when pgrp is not tool's.  This is the case of haven't
       * completed ttysw_fork yet (or even tried to do it yet).
       */
+ #ifndef __linux
      if (getpgrp(0) != pgrp)
+ #else
+     if (getpgrp() != pgrp)
+ #endif
  	/*
  	 * killpg could return -1 with errno == ESRCH but this is OK.
  	 */
***************
*** 368,373 ****
--- 372,381 ----
  	(void) notify_set_itimer_func((Notify_client) (TTY_PUBLIC(ttysw)),
  				      ttysw_itimer_expired,
  				ITIMER_REAL, &ttysw_itimerval, ITIMER_NULL);
+ #ifdef DEBUG
+    printf("ttysw_reset_conditions() waiting_for_pty_output=%d, waiting_for_pty_input=%d\n",
+ 	ttysw_waiting_for_pty_output, ttysw_waiting_for_pty_input);
+ #endif
  }
  
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/tty_stty.c xview3.2/lib//libxview/ttysw/tty_stty.c
*** xvorig/xview3.2/lib//libxview/ttysw/tty_stty.c	Tue Jun 29 08:17:23 1993
--- xview3.2/lib//libxview/ttysw/tty_stty.c	Sun Jul 11 13:26:10 1993
***************
*** 122,127 ****
--- 122,128 ----
       * often have a value of \0.
       */
      strcpy( str, WE_TTYPARMS_E );
+ #ifndef __linux
      (void) sprintf(str + strlen( str ),
  		"%ld,%ld,%ld,%ld,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd",
  		tp->c_iflag, tp->c_oflag, tp->c_cflag, tp->c_lflag,
***************
*** 129,134 ****
--- 130,145 ----
  		tp->c_cc[4],  tp->c_cc[5],  tp->c_cc[6],  tp->c_cc[7],
  		tp->c_cc[8],  tp->c_cc[9],  tp->c_cc[10], tp->c_cc[11],
  		tp->c_cc[12], tp->c_cc[13], tp->c_cc[14], tp->c_cc[15]);
+ #else /* __linux */
+     (void) sprintf(str + strlen( str ),
+ 		"%ld,%ld,%ld,%ld,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd,%hd",
+ 		tp->c_iflag, tp->c_oflag, tp->c_cflag, tp->c_lflag,
+ 		tp->c_cc[0],  tp->c_cc[1],  tp->c_cc[2],  tp->c_cc[3],
+ 		tp->c_cc[4],  tp->c_cc[5],  tp->c_cc[6],  tp->c_cc[7],
+ 		tp->c_cc[8],  tp->c_cc[9],  tp->c_cc[10], tp->c_cc[11],
+ 		tp->c_cc[12], tp->c_cc[13], tp->c_cc[14], tp->c_cc[15],
+ 		tp->c_cc[16]);
+ #endif
      (void) putenv(str);
  }
  
diff -c -r --new-file xvorig/xview3.2/lib//libxview/ttysw/ttyansi.c xview3.2/lib//libxview/ttysw/ttyansi.c
*** xvorig/xview3.2/lib//libxview/ttysw/ttyansi.c	Tue Jun 29 08:17:21 1993
--- xview3.2/lib//libxview/ttysw/ttyansi.c	Sat Aug 21 19:44:31 1993
***************
*** 48,54 ****
--- 48,58 ----
  #define ERROR_RETURN(val)	return(val);
  #endif				/* DEBUG */
  
+ #ifndef __linux
  #define notcontrol(c)	(((c&0177) >= ' ') && (c != '\177'))
+ #else
+ #define notcontrol(c)	((c >= ' ') && (c != '\177'))
+ #endif
  
  /* Logical state of window */
  int             curscol;	/* cursor column */
diff -c -r --new-file xvorig/xview3.2/lib//libxview/win/win_input.c xview3.2/lib//libxview/win/win_input.c
*** xvorig/xview3.2/lib//libxview/win/win_input.c	Tue Jun 29 08:15:41 1993
--- xview3.2/lib//libxview/win/win_input.c	Wed Aug  4 02:57:00 1993
***************
*** 77,83 ****
--- 77,85 ----
  Xv_private int  win_handle_menu_accel();
  /* ACC_XVIEW */
  
+ #ifndef __linux
  FILE           *fopen(), *fexp;
+ #endif
  
  Xv_object       xview_x_input_readevent();
  extern          input_imnull();
diff -c -r --new-file xvorig/xview3.2/lib//libxview/wmgr/wmgr_menu.c xview3.2/lib//libxview/wmgr/wmgr_menu.c
*** xvorig/xview3.2/lib//libxview/wmgr/wmgr_menu.c	Tue Jun 29 08:17:24 1993
--- xview3.2/lib//libxview/wmgr/wmgr_menu.c	Sun Jul 11 17:47:49 1993
***************
*** 14,21 ****
  #include <stdio.h>
  #include <ctype.h>
  #include <string.h>
! #ifdef sparc
! #ifdef SVR4
  #include <unistd.h>
  #else
  #include <vfork.h>
--- 14,21 ----
  #include <stdio.h>
  #include <ctype.h>
  #include <string.h>
! #if defined(sparc) || defined(__linux)
! #if defined(SVR4) || defined(__linux)
  #include <unistd.h>
  #else
  #include <vfork.h>
